{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Resolve Schema\n",
    "\n",
    "**input**\n",
    "```\n",
    "+schema_path: str\n",
    "```\n",
    "\n",
    "**methods**\n",
    "```\n",
    "+read_json(schema_path: str): dict\n",
    "+split_json(schema: dict): list\n",
    "+resolver(entity1: dict, entity2: dict): dict\n",
    "+resolve_defs(terms: dict, defs: dict) : dict\n",
    "+ node_order(schema: dict): list\n",
    "+resolve_nodes(nodeList: list, splitJsonList: list): list\n",
    "+recombine_nodes(resolvedList: list) : dict\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from collections import defaultdict, deque\n",
    "\n",
    "class ResolveSchema:\n",
    "    \n",
    "    def __init__(self, schema_path: str):\n",
    "        self.schema_path = schema_path\n",
    "        self.schema = self.read_json(self.schema_path)\n",
    "        self.nodes = self.get_nodes()\n",
    "        self.node_pairs = self.get_all_node_pairs()\n",
    "        self.node_order = self.get_node_order(edges=self.node_pairs)\n",
    "        \n",
    "        \n",
    "    def read_json(self, path: str):\n",
    "        with open(path) as f:\n",
    "            return json.load(f)\n",
    "    \n",
    "    \n",
    "    def get_nodes(self):\n",
    "        nodes = list(self.schema.keys())\n",
    "        return nodes\n",
    "    \n",
    "    \n",
    "    def get_node_link(self, node_name: str):\n",
    "        links = self.schema[node_name]['links']\n",
    "        node_id = self.schema[node_name]['id']\n",
    "        if 'subgroup' in links[0]:\n",
    "            return node_id, links[0]['subgroup']\n",
    "        else:\n",
    "            return node_id, links\n",
    "\n",
    "    def find_upstream_downstream(self, node_name: str):\n",
    "        \"\"\"Takes a node name and returns the upstream and downstream nodes\"\"\"\n",
    "        \n",
    "        node_id, links = self.get_node_link(node_name)\n",
    "        \n",
    "        # Ensure links is a list\n",
    "        if isinstance(links, dict):\n",
    "            links = [links]\n",
    "\n",
    "        results = []\n",
    "\n",
    "        for link in links:\n",
    "            target_type = link.get(\"target_type\")\n",
    "            \n",
    "            if not node_id or not target_type:\n",
    "                print(\"Missing essential keys in link:\", link)\n",
    "                results.append((None, None))\n",
    "                continue\n",
    "\n",
    "            results.append((target_type, node_id))\n",
    "\n",
    "        return results\n",
    "\n",
    "    def get_all_node_pairs(self, excluded_nodes=[\"_definitions.yaml\", \"_terms.yaml\", \"_settings.yaml\", \"program.yaml\"]):\n",
    "        node_pairs = []\n",
    "        for node in self.nodes:\n",
    "            if not node in excluded_nodes:\n",
    "                node_pairs.extend(self.find_upstream_downstream(node))\n",
    "            else:\n",
    "                continue\n",
    "        return node_pairs\n",
    "    \n",
    "    def get_node_order(self, edges: list):\n",
    "        # Build graph representation\n",
    "        graph = defaultdict(list)\n",
    "        in_degree = defaultdict(int)\n",
    "\n",
    "        for upstream, downstream in edges:\n",
    "            graph[upstream].append(downstream)\n",
    "            in_degree[downstream] += 1\n",
    "            if upstream not in in_degree:\n",
    "                in_degree[upstream] = 0\n",
    "\n",
    "        # Perform Topological Sorting (Kahn's Algorithm)\n",
    "        sorted_order = []\n",
    "        zero_in_degree = deque([node for node in in_degree if in_degree[node] == 0])\n",
    "\n",
    "        while zero_in_degree:\n",
    "            node = zero_in_degree.popleft()\n",
    "            sorted_order.append(node)\n",
    "            \n",
    "            for neighbor in graph[node]:\n",
    "                in_degree[neighbor] -= 1\n",
    "                if in_degree[neighbor] == 0:\n",
    "                    zero_in_degree.append(neighbor)\n",
    "\n",
    "        # Ensure core_metadata_collection is last\n",
    "        sorted_order.remove(\"core_metadata_collection\")\n",
    "        sorted_order.append(\"core_metadata_collection\")\n",
    "\n",
    "        return sorted_order\n",
    "        \n",
    "    # def split_json(self):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'demographic.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'clinical',\n",
       "  'description': 'Data for the characterization of the patient by means of segementing the population (e.g. characterization by age, sex, or race).',\n",
       "  'id': 'demographic',\n",
       "  'links': [{'backref': 'demographics',\n",
       "    'label': 'describes',\n",
       "    'multiplicity': 'one_to_one',\n",
       "    'name': 'subjects',\n",
       "    'required': True,\n",
       "    'target_type': 'subject'}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'abs_cluster': {'description': 'Australian bureau of statistics: Name of the suburb where the participant lives',\n",
       "    'type': 'string'},\n",
       "   'abs_state': {'description': 'Australian bureau of statistics: Name of the state where participant lives',\n",
       "    'type': 'string'},\n",
       "   'abs_weight_2000': {'description': 'Australian bureau of statistics: Population weights from 2000',\n",
       "    'type': ['number', 'null']},\n",
       "   'baseline_age': {'description': 'Age at baseline (examination/enrolment/visit 1) in years to one decimal place.',\n",
       "    'type': ['number', 'null']},\n",
       "   'baseline_date': {'description': 'Date of baseline examination/recruitment/visit 1. In the format [YYYY-MM-DD]',\n",
       "    'pattern': '^\\\\d{4}-\\\\d{2}-\\\\d{2}',\n",
       "    'type': 'string'},\n",
       "   'bmi_baseline': {'description': 'Body mass index calculated at baseline. (kg /(m^2))',\n",
       "    'termDef': [{'source': 'efo',\n",
       "      'term': 'body mass index',\n",
       "      'term_id': 'EFO:0004340'}],\n",
       "    'type': ['number', 'null']},\n",
       "   'education': {'description': 'Highest level of education at baseline. low = ABS HEAP [6-9] no education all the way to secondary education, medium = ABS HEAP [4-5] TAFE, diploma, or certificate, high = ABS HEAP [1-3] University qualification |  https://www.abs.gov.au/census/guide-census-data/census-dictionary/2021/variables-topic/education-and-training/level-highest-educational-attainment-heap',\n",
       "    'enum': ['low', 'medium', 'high']},\n",
       "   'education_abs_heap': {'description': 'Highest level of educational attainment based on ABS HEAP 2021 : https://www.abs.gov.au/census/guide-census-data/census-dictionary/2021/variables-topic/education-and-training/level-highest-educational-attainment-heap',\n",
       "    'enum': ['HEAP-1',\n",
       "     'HEAP-2',\n",
       "     'HEAP-3',\n",
       "     'HEAP-4',\n",
       "     'HEAP-5',\n",
       "     'HEAP-6',\n",
       "     'HEAP-7',\n",
       "     'HEAP-8',\n",
       "     'HEAP-9']},\n",
       "   'height_baseline_measured': {'description': 'Measured body height at baseline (m)',\n",
       "    'type': ['number', 'null']},\n",
       "   'height_baseline_self_reported': {'description': 'Self reported body height at baseline (m)',\n",
       "    'type': ['number', 'null']},\n",
       "   'month_birth': {'description': 'Month of birth at baseline (MM)',\n",
       "    'type': 'integer'},\n",
       "   'postcode': {'description': 'The postcode is used to calculate the Socio-Economic Indexes for Areas (SEIFA) quintile. It may be manually adjusted at the clinicianâ€™s discretion to better reflect the socioeconomic status of individual patients.',\n",
       "    'pattern': '^\\\\d{4}$',\n",
       "    'type': 'string'},\n",
       "   'sex': {'description': 'Self-reported biological sex of the participant',\n",
       "    'enum': ['male', 'female']},\n",
       "   'subjects': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'weight_baseline_measured': {'description': 'Measured body weight at baseline (kg)',\n",
       "    'type': ['number', 'null']},\n",
       "   'weight_baseline_self_reported': {'description': 'Self reported body weight at baseline (kg)',\n",
       "    'type': ['number', 'null']},\n",
       "   'year_birth': {'description': 'Year of birth at baseline (YYYY)',\n",
       "    'type': 'integer'}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'subjects',\n",
       "   'baseline_age',\n",
       "   'baseline_date',\n",
       "   'sex'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Demographic',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'project.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'administrative',\n",
       "  'constraints': None,\n",
       "  'description': 'The study the data is coming from',\n",
       "  'id': 'project',\n",
       "  'links': [{'backref': 'projects',\n",
       "    'label': 'member_of',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'name': 'programs',\n",
       "    'required': True,\n",
       "    'target_type': 'program'}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'authz': {'description': 'The authz string (example, [\"/programs/Program.name/projects/Project.code\"] ).',\n",
       "    'items': {'type': 'string'},\n",
       "    'type': 'array'},\n",
       "   'availability_mechanism': {'description': 'Mechanism by which the project will be made avilable.',\n",
       "    'type': 'string'},\n",
       "   'availability_type': {'description': 'Is the project open or restricted?',\n",
       "    'enum': ['Open', 'Restricted']},\n",
       "   'code': {'description': 'Unique identifier for the project.',\n",
       "    'type': 'string'},\n",
       "   'data_access_url': {'description': 'A URL link that takes you to a page that explains how to gain access to the project',\n",
       "    'type': 'string'},\n",
       "   'date_collected': {'description': 'The date or date range in which the project data was collected.',\n",
       "    'type': 'string'},\n",
       "   'dbgap_accession_number': {'description': 'The dbgap accession number provided for the project. (CMG)',\n",
       "    'type': 'string'},\n",
       "   'full_name': {'description': 'The full name for the project.',\n",
       "    'type': 'string'},\n",
       "   'id': {'$ref': '_definitions.yaml#/UUID',\n",
       "    'description': 'UUID for the project.',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'intended_release_date': {'description': \"Tracks a Project's intended release date.\",\n",
       "    'format': 'date-time',\n",
       "    'type': 'string'},\n",
       "   'name': {'description': 'Display name/brief description for the project.',\n",
       "    'type': 'string'},\n",
       "   'programs': {'$ref': '_definitions.yaml#/to_one',\n",
       "    'description': 'Indicates that the Study is logically part of the indicated Study.\\n'},\n",
       "   'project_description': {'description': 'The description for the project.',\n",
       "    'type': 'string'},\n",
       "   'project_registration': {'description': 'External source from which the identifier included in study_id originates',\n",
       "    'type': 'string'},\n",
       "   'releasable': {'default': False,\n",
       "    'description': 'A project can only be released by the user when `releasable` is true.\\n',\n",
       "    'type': 'boolean'},\n",
       "   'released': {'default': False,\n",
       "    'description': 'To release a project is to tell the GDC to include all submitted\\nentities in the next GDC index.\\n',\n",
       "    'type': 'boolean'},\n",
       "   'short_name': {'description': 'The shorthand name for the project.',\n",
       "    'type': 'string'},\n",
       "   'state': {'default': 'open',\n",
       "    'description': 'The possible states a project can be in.  All but `open` are\\nequivalent to some type of locked state.\\n',\n",
       "    'enum': ['open', 'review', 'submitted', 'processing', 'closed', 'legacy']},\n",
       "   'support_id': {'description': 'The ID of the source providing support/grant resources.',\n",
       "    'type': 'string'},\n",
       "   'support_source': {'description': 'The name of source providing support/grant resources.',\n",
       "    'type': 'string'},\n",
       "   'type': {'type': 'string'}},\n",
       "  'required': ['code', 'name', 'programs', 'dbgap_accession_number'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'state',\n",
       "   'released',\n",
       "   'releasable',\n",
       "   'intended_release_date'],\n",
       "  'title': 'Project',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['code']],\n",
       "  'validators': None},\n",
       " 'serum_marker_assay.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'experimental_methods',\n",
       "  'description': 'Details about the methods that produced the serum marker file',\n",
       "  'id': 'serum_marker_assay',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'serum_marker_assays',\n",
       "      'label': 'produces',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'samples',\n",
       "      'required': False,\n",
       "      'target_type': 'sample'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'assay_description': {'description': 'Description of the method used to assay the markers',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'samples': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'serum_markers': {'description': 'A list of the serum markers that were measured',\n",
       "    'items': {'enum': ['TC',\n",
       "      'HDL',\n",
       "      'TGL',\n",
       "      'BNP',\n",
       "      'GAL3',\n",
       "      'Trop',\n",
       "      'HCO3',\n",
       "      'Cr',\n",
       "      'CRP',\n",
       "      'GLU',\n",
       "      'K',\n",
       "      'LP(A)',\n",
       "      'UREA',\n",
       "      'Omega-3'],\n",
       "     'enumDef': [{'enumeration': 'TC',\n",
       "       'source': 'NCIT',\n",
       "       'term_id': 'NCIT:C61032'},\n",
       "      {'enumeration': 'HDL', 'source': 'NCIT', 'term_id': 'NCIT:C25186'},\n",
       "      {'enumeration': 'TGL', 'source': 'NCIT', 'term_id': 'NCIT:C61037'},\n",
       "      {'enumeration': 'BNP', 'source': 'NCIT', 'term_id': 'NCIT:C74735'},\n",
       "      {'enumeration': 'GAL3', 'source': 'NCIT', 'term_id': 'NCIT:C172493'},\n",
       "      {'enumeration': 'Trop', 'source': 'NCIT', 'term_id': 'NCIT:C111327'},\n",
       "      {'enumeration': 'HCO3', 'source': 'NCIT', 'term_id': 'NCIT:C61022'},\n",
       "      {'enumeration': 'Cr', 'source': 'NCIT', 'term_id': 'NCIT:C61023'},\n",
       "      {'enumeration': 'CRP', 'source': 'NCIT', 'term_id': 'NCIT:C64548'},\n",
       "      {'enumeration': 'GLU', 'source': 'NCIT', 'term_id': 'NCIT:C61027'},\n",
       "      {'enumeration': 'K', 'source': 'NCIT', 'term_id': 'NCIT:C61030'},\n",
       "      {'enumeration': 'LP(A)', 'source': 'NCIT', 'term_id': 'NCIT:C82022'},\n",
       "      {'source': 'NCIT', 'term_id': 'NCIT:C64809'},\n",
       "      {'enumeration': 'UREA', 'source': 'NCIT', 'term_id': 'NCIT:C64815'},\n",
       "      {'enumeration': 'Omega-3', 'source': 'EFO', 'term_id': 'EFO:0010119'}]},\n",
       "    'type': 'array'}},\n",
       "  'required': ['type', 'submitter_id', 'assay_description'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Serum Marker Assay',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'alignment_workflow.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'analysis',\n",
       "  'description': 'Metadata for the alignment pipeline used to align reads during genomic harmonisation',\n",
       "  'id': 'alignment_workflow',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'alignment_workflows',\n",
       "      'label': 'part_of',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'unaligned_reads_files',\n",
       "      'required': False,\n",
       "      'target_type': 'unaligned_reads_file'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'unaligned_reads_files': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'workflow_end_datetime': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]',\n",
       "    'pattern': '^\\\\d{4}-\\\\d{2}-\\\\d{2}T\\\\d{2}:\\\\d{2}:\\\\d{2}(?:\\\\.\\\\d+)?(?:Z|[+-]\\\\d{2}:\\\\d{2})$',\n",
       "    'type': 'string'},\n",
       "   'workflow_link': {'description': 'Link to Github hash for the CWL workflow used.',\n",
       "    'type': 'string'},\n",
       "   'workflow_start_datetime': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]',\n",
       "    'pattern': '^\\\\d{4}-\\\\d{2}-\\\\d{2}T\\\\d{2}:\\\\d{2}:\\\\d{2}(?:\\\\.\\\\d+)?(?:Z|[+-]\\\\d{2}:\\\\d{2})$',\n",
       "    'type': 'string'},\n",
       "   'workflow_type': {'description': 'Type of read aligner used',\n",
       "    'enum': ['STAR', 'BWA-aln', 'BWA-mem', 'spinnaker'],\n",
       "    'enumDef': [{'enumeration': 'STAR', 'source': 'BioData Catalyst DD'},\n",
       "     {'enumeration': 'BWA-aln', 'source': 'BioData Catalyst DD'},\n",
       "     {'enumeration': 'BWA-mem', 'source': 'BioData Catalyst DD'},\n",
       "     {'enumeration': 'spinnaker', 'source': 'BioData Catalyst DD'}]},\n",
       "   'workflow_version': {'description': 'Version of the workflow used',\n",
       "    'type': 'string'}},\n",
       "  'required': ['type', 'submitter_id', 'workflow_type'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Alignment Workflow',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'imaging_file.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'data_file',\n",
       "  'description': 'Data file containing medical Images',\n",
       "  'id': 'imaging_file',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'imaging_files',\n",
       "      'label': 'data_from',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'subjects',\n",
       "      'required': False,\n",
       "      'target_type': 'subject'},\n",
       "     {'backref': 'imaging_files',\n",
       "      'label': 'data_from',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'core_metadata_collections',\n",
       "      'required': False,\n",
       "      'target_type': 'core_metadata_collection'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/data_file_properties',\n",
       "   'acquisition_type': {'description': 'The method used to acquire the images, such as SEQUENCED, SPIRAL, CONSTANT_ANGLE, STATIONARY, or FREE.',\n",
       "    'enum': ['SEQUENCED', 'SPIRAL', 'CONSTANT_ANGLE', 'STATIONARY', 'FREE']},\n",
       "   'contrast_bolus_agent': {'description': 'The type of contrast agent used during the imaging procedure, such as iodine-based agents.',\n",
       "    'type': 'string'},\n",
       "   'convolution_kernel': {'description': 'The type of convolution kernel used in image reconstruction, which affects image sharpness and noise.',\n",
       "    'type': 'string'},\n",
       "   'core_metadata_collections': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'data_category': {'description': 'The broad category of the imaging data, such as CT, MRI, X-ray, Ultrasound, or PET.',\n",
       "    'enum': ['CT', 'MRI', 'X-Ray', 'Ultrasound', 'PET']},\n",
       "   'data_format': {'description': 'The specific format of the data files, such as DCM for DICOM files or nii.gz for NIfTI files.',\n",
       "    'enum': ['dcm', 'nii']},\n",
       "   'data_type': {'description': 'The type of data file, such as DICOM or NIfTI, which indicates the format used for storing medical images.',\n",
       "    'enum': ['DICOM', 'NifTI']},\n",
       "   'exposure_modulation_type': {'description': 'The type of exposure modulation used to adjust the radiation dose during imaging, such as AUTO.',\n",
       "    'type': 'string'},\n",
       "   'image_data_modification_method': {'description': 'The type of modification performed on the image data, such as removal of alphanumeric characters, defacing, or skull-stripping.',\n",
       "    'enum': ['Removal of alphanumeric characters',\n",
       "     'Defacing',\n",
       "     'Skull-stripping']},\n",
       "   'image_data_modification_name': {'description': 'A classification describing specific methods used to modify the image data, such as software tools or processes.',\n",
       "    'type': 'integer'},\n",
       "   'image_data_modified': {'description': 'Indicates whether any post-clinical modification of the image pixel data has been performed.',\n",
       "    'type': 'boolean'},\n",
       "   'lossy_image_compression': {'description': 'Indicates whether the image was compressed using a lossy algorithm, which can reduce file size but may affect image quality.',\n",
       "    'type': 'boolean'},\n",
       "   'manufacturer': {'description': 'The name of the manufacturer of the imaging equipment, such as GE Healthcare or Siemens.',\n",
       "    'type': 'string'},\n",
       "   'manufacturer_model_name': {'description': 'The model name of the imaging equipment, such as Revolution CT or MAGNETOM Skyra.',\n",
       "    'type': 'string'},\n",
       "   'modality': {'description': 'The type of imaging modality used, such as CT, MRI, X-ray, Ultrasound, or PET.',\n",
       "    'type': 'string'},\n",
       "   'object_id': {'description': 'The unique identifier for the imaging file in the data index service.',\n",
       "    'type': 'string'},\n",
       "   'patient_position': {'description': 'The position of the patient during the imaging procedure, such as Head First Supine (HFS) or Feet First Prone (FFP).',\n",
       "    'type': 'string'},\n",
       "   'pixel_spacing': {'description': 'The distance between the centers of adjacent pixels in the image, measured in millimeters.',\n",
       "    'type': ['number', 'null']},\n",
       "   'series_description': {'description': 'A description of the imaging series, such as \"Chest CT\" or \"Cardiac MRI\".',\n",
       "    'type': 'string'},\n",
       "   'series_uid': {'description': 'The unique identifier for the imaging series, assigned by the imaging equipment.',\n",
       "    'type': 'string'},\n",
       "   'slice_thickness': {'description': 'The thickness of each slice in the imaging series, measured in millimeters.',\n",
       "    'type': ['number', 'null']},\n",
       "   'spacing_between_slices': {'description': 'The distance between the centers of adjacent slices in the imaging series, measured in millimeters.',\n",
       "    'type': ['number', 'null']},\n",
       "   'subjects': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'data_type',\n",
       "   'data_format',\n",
       "   'data_category',\n",
       "   'acquisition_type'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Imaging File',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'lipidomics_assay.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'experimental_methods',\n",
       "  'description': 'Details about the methods that produced the lipidomics file',\n",
       "  'id': 'lipidomics_assay',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'lipidomics_assays',\n",
       "      'label': 'produces',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'samples',\n",
       "      'required': False,\n",
       "      'target_type': 'sample'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'assay_description': {'description': 'A description of the type of assay used.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'instrument_type': {'description': 'Type of mass spec the instrument used for lipidome measurements',\n",
       "    'enum': ['LC-MS',\n",
       "     'MS/MS',\n",
       "     'MALDI MS',\n",
       "     'Ion mobility MS',\n",
       "     'Tandem MS',\n",
       "     'High mass accuracy MS',\n",
       "     'multidimensional MS',\n",
       "     'DESI MS',\n",
       "     'SIMS']},\n",
       "   'lipid_names': {'description': 'A list of all lipids assayed in this method',\n",
       "    'items': {'enum': ['Sph(d16:1)',\n",
       "      'Sph(d17:1)',\n",
       "      'Sph(d18:1)',\n",
       "      'Sph(d18:2)',\n",
       "      'S1P(d16:1)',\n",
       "      'S1P(d18:0)',\n",
       "      'S1P(d18:1)',\n",
       "      'S1P(d18:2)',\n",
       "      'dhCer(d18:0/16:0)',\n",
       "      'dhCer(d18:0/18:0)',\n",
       "      'dhCer(d18:0/20:0)',\n",
       "      'dhCer(d18:0/22:0)',\n",
       "      'dhCer(d18:0/24:0)',\n",
       "      'dhCer(d18:0/24:1)',\n",
       "      'Cer(d16:1/16:0)',\n",
       "      'Cer(d16:1/18:0)',\n",
       "      'Cer(d16:1/20:0)',\n",
       "      'Cer(d16:1/22:0)',\n",
       "      'Cer(d16:1/23:0)',\n",
       "      'Cer(d16:1/24:0)',\n",
       "      'Cer(d16:1/24:1)',\n",
       "      'Cer(d17:1/16:0)',\n",
       "      'Cer(d17:1/18:0)',\n",
       "      'Cer(d17:1/20:0)',\n",
       "      'Cer(d17:1/22:0)',\n",
       "      'Cer(d17:1/23:0)',\n",
       "      'Cer(d17:1/24:0)',\n",
       "      'Cer(d17:1/24:1)',\n",
       "      'Cer(d18:1/14:0)',\n",
       "      'Cer(d18:1/16:0)',\n",
       "      'Cer(d18:1/17:0)',\n",
       "      'Cer(d18:1/18:0)',\n",
       "      'Cer(d18:1/19:0)',\n",
       "      'Cer(d18:1/20:0)',\n",
       "      'Cer(d18:1/21:0)',\n",
       "      'Cer(d18:1/22:0)',\n",
       "      'Cer(d18:1/23:0)',\n",
       "      'Cer(d18:1/24:0)',\n",
       "      'Cer(d18:1/24:1)',\n",
       "      'Cer(d18:1/26:0)',\n",
       "      'Cer(d18:2/14:0)',\n",
       "      'Cer(d18:2/16:0)',\n",
       "      'Cer(d18:2/17:0)',\n",
       "      'Cer(d18:2/18:0)',\n",
       "      'Cer(d18:2/20:0)',\n",
       "      'Cer(d18:2/21:0)',\n",
       "      'Cer(d18:2/22:0)',\n",
       "      'Cer(d18:2/23:0)',\n",
       "      'Cer(d18:2/24:0)',\n",
       "      'Cer(d18:2/24:1)',\n",
       "      'Cer(d18:2/26:0)',\n",
       "      'Cer(d19:1/16:0)',\n",
       "      'Cer(d19:1/18:0)',\n",
       "      'Cer(d19:1/20:0)',\n",
       "      'Cer(d19:1/22:0)',\n",
       "      'Cer(d19:1/23:0)',\n",
       "      'Cer(d19:1/24:0)',\n",
       "      'Cer(d19:1/24:1)',\n",
       "      'Cer(d19:1/26:0)',\n",
       "      'Cer(d20:1/22:0)',\n",
       "      'Cer(d20:1/23:0)',\n",
       "      'Cer(d20:1/24:0)',\n",
       "      'Cer(d20:1/24:1)',\n",
       "      'Cer(d20:1/26:0)',\n",
       "      'Cer(m18:0/20:0)',\n",
       "      'Cer(m18:0/22:0)',\n",
       "      'Cer(m18:0/23:0)',\n",
       "      'Cer(m18:0/24:0)',\n",
       "      'Cer(m18:0/24:1)',\n",
       "      'Cer(m18:1/18:0)',\n",
       "      'Cer(m18:1/20:0)',\n",
       "      'Cer(m18:1/22:0)',\n",
       "      'Cer(m18:1/23:0)',\n",
       "      'Cer(m18:1/24:0)',\n",
       "      'Cer(m18:1/24:1)',\n",
       "      'HexCer(d16:1/18:0)',\n",
       "      'HexCer(d16:1/20:0)',\n",
       "      'HexCer(d16:1/22:0)',\n",
       "      'HexCer(d16:1/24:0)',\n",
       "      'HexCer(d18:1/16:0)',\n",
       "      'HexCer(d18:1/18:0)',\n",
       "      'HexCer(d18:1/20:0)',\n",
       "      'HexCer(d18:1/22:0)',\n",
       "      'HexCer(d18:1/24:0)',\n",
       "      'HexCer(d18:1/24:1)',\n",
       "      'HexCer(d18:2/18:0)',\n",
       "      'HexCer(d18:2/20:0)',\n",
       "      'HexCer(d18:2/22:0)',\n",
       "      'HexCer(d18:2/24:0)',\n",
       "      'Hex2Cer(d16:1/16:0)',\n",
       "      'Hex2Cer(d16:1/24:1)',\n",
       "      'Hex2Cer(d18:1/16:0)',\n",
       "      'Hex2Cer(d18:1/18:0)',\n",
       "      'Hex2Cer(d18:1/20:0)',\n",
       "      'Hex2Cer(d18:1/22:0)',\n",
       "      'Hex2Cer(d18:1/24:0)',\n",
       "      'Hex2Cer(d18:1/24:1)',\n",
       "      'Hex2Cer(d18:2/16:0)',\n",
       "      'Hex2Cer(d18:2/24:1)',\n",
       "      'Hex3Cer(d18:1/16:0)',\n",
       "      'Hex3Cer(d18:1/18:0)',\n",
       "      'Hex3Cer(d18:1/20:0)',\n",
       "      'Hex3Cer(d18:1/22:0)',\n",
       "      'Hex3Cer(d18:1/24:0)',\n",
       "      'Hex3Cer(d18:1/24:1)',\n",
       "      'GM3(d18:1/16:0)',\n",
       "      'GM3(d18:1/18:0)',\n",
       "      'GM3(d18:1/20:0)',\n",
       "      'GM3(d18:1/22:0)',\n",
       "      'GM3(d18:1/24:0)',\n",
       "      'GM3(d18:1/24:1)',\n",
       "      'GM3(d18:2/24:1)',\n",
       "      'GM3(d20:1/18:0)',\n",
       "      'GM1(d18:1/16:0)',\n",
       "      'Sulfatide (d18:1:/16:0(OH))',\n",
       "      'Sulfatide (d18:1:/16:0)',\n",
       "      'Sulfatide (d18:1:/18:0(OH))',\n",
       "      'Sulfatide (d18:1:/18:0)',\n",
       "      'Sulfatide (d18:1:/24:0(OH))',\n",
       "      'Sulfatide (d18:1:/24:0)',\n",
       "      'Sulfatide (d18:1:/24:1(OH))',\n",
       "      'Sulfatide (d18:1:/24:1)',\n",
       "      'Sulfatide (d18:2:/18:0(OH))',\n",
       "      'Sulfatide (d18:2:/18:0)',\n",
       "      'Cer1P(d18:1/16:0)',\n",
       "      'SM(34:3)',\n",
       "      'SM(35:2) (b)',\n",
       "      'SM(37:1)',\n",
       "      'SM(37:2)',\n",
       "      'SM(38:3) (a)',\n",
       "      'SM(38:3) (b)',\n",
       "      'SM(40:3) (a)',\n",
       "      'SM(40:3) (b)',\n",
       "      'SM(41:0)',\n",
       "      'SM(41:1) (a)',\n",
       "      'SM(43:1)',\n",
       "      'SM(43:2) (b)',\n",
       "      'SM(43:2) (c)',\n",
       "      'SM(44:1)',\n",
       "      'SM(44:2)',\n",
       "      'SM(44:3)',\n",
       "      'SM(d16:1/19:0)',\n",
       "      'SM(d16:1/23:0)/SM(d17:1/22:0)',\n",
       "      'SM(d16:1/24:1)',\n",
       "      'SM(d17:1/14:0)',\n",
       "      'SM(d17:1/16:0)',\n",
       "      'SM(d17:1/24:1)',\n",
       "      'SM(d18:0/14:0)',\n",
       "      'SM(d18:0/16:0)',\n",
       "      'SM(d18:0/22:0)',\n",
       "      'SM(d18:1/14:0)/SM(d16:1/16:0)',\n",
       "      'SM(d18:1/16:0)',\n",
       "      'SM(d18:1/17:0)/SM(d17:1/18:0)',\n",
       "      'SM(d18:1/18:0)/SM(d16:1/20:0)',\n",
       "      'SM(d18:1/20:0)/SM(d16:1/22:0)',\n",
       "      'SM(d18:1/22:0)/SM(d16:1/24:0)',\n",
       "      'SM(d18:1/23:0)/SM(d17:1/24:0)',\n",
       "      'SM(d18:1/24:0)',\n",
       "      'SM(d18:1/24:1)',\n",
       "      'SM(d18:2/14:0)',\n",
       "      'SM(d18:2/16:0)',\n",
       "      'SM(d18:2/17:0)',\n",
       "      'SM(d18:2/18:0)',\n",
       "      'SM(d18:2/18:1)',\n",
       "      'SM(d18:2/20:0)',\n",
       "      'SM(d18:2/22:0)',\n",
       "      'SM(d18:2/23:0)',\n",
       "      'SM(d18:2/24:0)',\n",
       "      'SM(d19:1/24:1)',\n",
       "      'PC (39:5) (a)',\n",
       "      'PC (39:5) (b)',\n",
       "      'PC(14:0_16:0)',\n",
       "      'PC(14:0_20:4)',\n",
       "      'PC(14:0_22:6)',\n",
       "      'PC(15-MHDA_18:1)',\n",
       "      'PC(15-MHDA_18:2)',\n",
       "      'PC(15-MHDA_20:4)',\n",
       "      'PC(15-MHDA_22:6)',\n",
       "      'PC(15:0_20:3)',\n",
       "      'PC(15:0_20:4)',\n",
       "      'PC(15:0_22:6)',\n",
       "      'PC(16:0_16:0)',\n",
       "      'PC(16:0_18:0)',\n",
       "      'PC(16:0_18:1)',\n",
       "      'PC(16:0_18:2)',\n",
       "      'PC(16:0_18:3) (a)',\n",
       "      'PC(16:0_18:3) (b)',\n",
       "      'PC(16:0_20:3) (a)',\n",
       "      'PC(16:0_20:3) (b)',\n",
       "      'PC(16:0_20:4)',\n",
       "      'PC(16:0_20:5)',\n",
       "      'PC(16:0_22:6)',\n",
       "      'PC(16:1_18:2)',\n",
       "      'PC(16:1_20:4)',\n",
       "      'PC(16:1_22:6)',\n",
       "      'PC(17:0_18:1)',\n",
       "      'PC(17:0_18:2)',\n",
       "      'PC(17:0_20:4)',\n",
       "      'PC(17:0_22:6)',\n",
       "      'PC(17:1_18:2)',\n",
       "      'PC(18:0_18:1)',\n",
       "      'PC(18:0_18:2)',\n",
       "      'PC(18:0_20:3)',\n",
       "      'PC(18:0_20:4)',\n",
       "      'PC(18:0_22:4)',\n",
       "      'PC(18:0_22:5) (n3)/PC(20:1_20:4)',\n",
       "      'PC(18:0_22:5) (n6)',\n",
       "      'PC(18:0_22:6)',\n",
       "      'PC(18:1_18:1)',\n",
       "      'PC(18:1_18:2)',\n",
       "      'PC(18:1_20:3)',\n",
       "      'PC(18:1_22:6) (a)',\n",
       "      'PC(18:1_22:6) (b)',\n",
       "      'PC(18:2_18:2)',\n",
       "      'PC(18:2_20:5)',\n",
       "      'PC(20:0_20:4)',\n",
       "      'PC(28:0)',\n",
       "      'PC(31:0) (a)',\n",
       "      'PC(31:0) (b)',\n",
       "      'PC(32:1)',\n",
       "      'PC(32:2)',\n",
       "      'PC(33:0) (a)',\n",
       "      'PC(33:0) (b)',\n",
       "      'PC(33:1)',\n",
       "      'PC(33:2)',\n",
       "      'PC(34:5)',\n",
       "      'PC(35:5)',\n",
       "      'PC(36:6) (a)',\n",
       "      'PC(38:2)',\n",
       "      'PC(38:4) (b)',\n",
       "      'PC(38:5) (a)',\n",
       "      'PC(38:5) (b)',\n",
       "      'PC(38:6) (a)',\n",
       "      'PC(38:7) (c)',\n",
       "      'PC(40:10)',\n",
       "      'PC(40:7) (a)',\n",
       "      'PC(40:8)',\n",
       "      'PC(44:12)',\n",
       "      'PC(O-16:0/16:0)',\n",
       "      'PC(O-16:0/20:3)',\n",
       "      'PC(O-16:0/20:4)',\n",
       "      'PC(O-16:0/22:6)',\n",
       "      'PC(O-18:0/18:1)',\n",
       "      'PC(O-18:0/18:2)',\n",
       "      'PC(O-18:0/20:4)',\n",
       "      'PC(O-18:0/22:6)',\n",
       "      'PC(O-18:1/18:1)',\n",
       "      'PC(O-18:1/18:2)',\n",
       "      'PC(O-32:1)',\n",
       "      'PC(O-32:2)',\n",
       "      'PC(O-34:1)',\n",
       "      'PC(O-34:2)',\n",
       "      'PC(O-34:4)',\n",
       "      'PC(O-35:4)',\n",
       "      'PC(O-36:0)',\n",
       "      'PC(O-36:5)',\n",
       "      'PC(O-38:5)',\n",
       "      'PC(O-40:5)',\n",
       "      'PC(O-40:7) (a)',\n",
       "      'PC(O-40:7) (b)',\n",
       "      'PC(P-15:0/20:4) (a)',\n",
       "      'PC(P-15:0/20:4) (b)',\n",
       "      'PC(P-16:0/14:0)',\n",
       "      'PC(P-16:0/16:0)',\n",
       "      'PC(P-16:0/16:1)',\n",
       "      'PC(P-16:0/18:0)',\n",
       "      'PC(P-16:0/18:1)',\n",
       "      'PC(P-16:0/18:2)',\n",
       "      'PC(P-16:0/18:3)',\n",
       "      'PC(P-16:0/20:4)',\n",
       "      'PC(P-16:0/20:5)',\n",
       "      'PC(P-16:0/22:6)',\n",
       "      'PC(P-17:0/20:4) (a)',\n",
       "      'PC(P-17:0/20:4) (b)',\n",
       "      'PC(P-18:0/18:2)',\n",
       "      'PC(P-18:0/20:4)',\n",
       "      'PC(P-18:0/22:5)',\n",
       "      'PC(P-18:0/22:6)',\n",
       "      'PC(P-18:1/18:1)',\n",
       "      'PC(P-20:0/20:4)',\n",
       "      'PC(P-35:2)(a)',\n",
       "      'PC(P-35:2)(b)',\n",
       "      'PC(P-36:3)',\n",
       "      'PC(P-38:5) (a)',\n",
       "      'PC(P-38:5) (b)',\n",
       "      'LPC(14:0) [sn1]',\n",
       "      'LPC(14:0) [sn2]',\n",
       "      'LPC(15-MHDA) [sn1] / LPC(17:0) [sn2]',\n",
       "      'LPC(15-MHDA) [sn1] [104_sn1]',\n",
       "      'LPC(15-MHDA) [sn2]',\n",
       "      'LPC(15:0) [sn1]',\n",
       "      'LPC(15:0) [sn2]',\n",
       "      'LPC(16:0) [sn1]',\n",
       "      'LPC(16:0) [sn2]',\n",
       "      'LPC(16:1) [sn1]',\n",
       "      'LPC(16:1) [sn2]',\n",
       "      'LPC(17:0) [sn1]',\n",
       "      'LPC(17:1 ) [sn2] (a)',\n",
       "      'LPC(17:1) (a) [sn1] [104_sn1]',\n",
       "      'LPC(17:1) [sn1] (a) / LPC(17:1) [sn2] (b)',\n",
       "      'LPC(17:1) [sn1] (b)',\n",
       "      'LPC(18:0) [sn1]',\n",
       "      'LPC(18:0) [sn2]',\n",
       "      'LPC(18:1) [sn1]',\n",
       "      'LPC(18:1) [sn2]',\n",
       "      'LPC(18:2) [sn1]',\n",
       "      'LPC(18:2) [sn2]',\n",
       "      'LPC(18:3) (a) [sn1] [104_sn1]',\n",
       "      'LPC(18:3) [sn1] (a)/LPC(18:3) [sn2] (b)',\n",
       "      'LPC(18:3) [sn1] (b)',\n",
       "      'LPC(18:3) [sn2] (a)',\n",
       "      'LPC(19:0) (a) [sn1] [104_sn1]',\n",
       "      'LPC(19:0) [sn1] (a) / LPC(19:0) [sn2] (b)',\n",
       "      'LPC(19:0) [sn1] (b)',\n",
       "      'LPC(19:0) [sn2] (a)',\n",
       "      'LPC(19:1) (a)',\n",
       "      'LPC(19:1) (b)',\n",
       "      'LPC(19:1) (c)',\n",
       "      'LPC(20:0) [sn1]',\n",
       "      'LPC(20:0) [sn2]',\n",
       "      'LPC(20:1) [sn1]',\n",
       "      'LPC(20:1) [sn2]',\n",
       "      'LPC(20:2) [sn1]',\n",
       "      'LPC(20:2) [sn2]',\n",
       "      'LPC(20:3) [sn1]',\n",
       "      'LPC(20:3) [sn2]',\n",
       "      'LPC(20:4) [sn1]',\n",
       "      'LPC(20:4) [sn2]',\n",
       "      'LPC(20:5) [sn1]',\n",
       "      'LPC(20:5) [sn2]',\n",
       "      'LPC(22:0) [sn1]',\n",
       "      'LPC(22:0) [sn2]',\n",
       "      'LPC(22:1) [sn1]',\n",
       "      'LPC(22:1) [sn2]',\n",
       "      'LPC(22:4) [sn1]',\n",
       "      'LPC(22:4) [sn2]',\n",
       "      'LPC(22:5) (n3) [sn1] [104_sn1]',\n",
       "      'LPC(22:5) [sn1] (n3)/LPC(22:5) [sn2] (n6)',\n",
       "      'LPC(22:5) [sn1] (n6)',\n",
       "      'LPC(22:5) [sn2] (n3)',\n",
       "      'LPC(22:6) [sn1]',\n",
       "      'LPC(22:6) [sn2]',\n",
       "      'LPC(24:0) [sn1]',\n",
       "      'LPC(24:0) [sn2]',\n",
       "      'LPC(26:0) [sn1]',\n",
       "      'LPC(26:0) [sn2]',\n",
       "      'PC(P-18:1/22:6)',\n",
       "      'LPC(O-16:0)',\n",
       "      'LPC(O-18:0)',\n",
       "      'LPC(O-18:1)',\n",
       "      'LPC(O-20:0)',\n",
       "      'LPC(O-20:1)',\n",
       "      'LPC(O-22:0)',\n",
       "      'LPC(O-22:1)',\n",
       "      'LPC(O-24:0)',\n",
       "      'LPC(O-24:1)',\n",
       "      'LPC(O-24:2)',\n",
       "      'LPC(P-16:0)',\n",
       "      'LPC(P-17:0)(a)',\n",
       "      'LPC(P-17:0)(b)',\n",
       "      'LPC(P-18:0)',\n",
       "      'LPC(P-18:1)',\n",
       "      'LPC(P-20:0)',\n",
       "      'PE(15-MHDA_18:1)',\n",
       "      'PE(15-MHDA_18:2)',\n",
       "      'PE(15-MHDA_20:4)',\n",
       "      'PE(15-MHDA_22:6)',\n",
       "      'PE(16:0_16:0)',\n",
       "      'PE(16:0_16:1)',\n",
       "      'PE(16:0_18:1)',\n",
       "      'PE(16:0_18:2)',\n",
       "      'PE(16:0_18:3) (a)',\n",
       "      'PE(16:0_18:3) (b)',\n",
       "      'PE(16:0_20:3)',\n",
       "      'PE(16:0_20:4)',\n",
       "      'PE(16:0_20:5)',\n",
       "      'PE(16:0_22:6)',\n",
       "      'PE(16:1_18:2)',\n",
       "      'PE(16:1_20:4)',\n",
       "      'PE(17:0_18:1)',\n",
       "      'PE(17:0_18:2)',\n",
       "      'PE(17:0_20:4)',\n",
       "      'PE(17:0_22:6)',\n",
       "      'PE(18:0_18:1)',\n",
       "      'PE(18:0_18:2)',\n",
       "      'PE(18:0_20:3) (a)',\n",
       "      'PE(18:0_20:3) (b)',\n",
       "      'PE(18:0_20:4)',\n",
       "      'PE(18:0_22:4)',\n",
       "      'PE(18:0_22:5) (n3)',\n",
       "      'PE(18:0_22:5) (n6)',\n",
       "      'PE(18:0_22:6)',\n",
       "      'PE(18:1_18:1)',\n",
       "      'PE(18:1_18:2)',\n",
       "      'PE(18:1_22:6) (a)',\n",
       "      'PE(18:1_22:6) (b)',\n",
       "      'PE(20:0_20:4)',\n",
       "      'PE(36:0)',\n",
       "      'PE(38:5) (a)',\n",
       "      'PE(38:5) (b)',\n",
       "      'PE(O-16:0/18:2)',\n",
       "      'PE(O-16:0/20:3)',\n",
       "      'PE(O-16:0/20:4)',\n",
       "      'PE(O-16:0/22:4)',\n",
       "      'PE(O-16:0/22:6)',\n",
       "      'PE(O-18:0/20:4)',\n",
       "      'PE(O-18:0/22:5) (a)',\n",
       "      'PE(O-18:0/22:6)',\n",
       "      'PE(O-18:1/18:2)',\n",
       "      'PE(O-18:1/22:6)',\n",
       "      'PE(O-34:1)',\n",
       "      'PE(O-36:5)',\n",
       "      'PE(O-38:5) (a)',\n",
       "      'PE(O-38:5) (b)',\n",
       "      'PE(P-15:0/20:4)',\n",
       "      'PE(P-15:0/22:6)(a)',\n",
       "      'PE(P-15:0/22:6)(b)',\n",
       "      'PE(P-16:0/18:1)',\n",
       "      'PE(P-16:0/18:2)',\n",
       "      'PE(P-16:0/18:3)',\n",
       "      'PE(P-16:0/20:3)(a)',\n",
       "      'PE(P-16:0/20:3)(b)',\n",
       "      'PE(P-16:0/20:4)',\n",
       "      'PE(P-16:0/20:5)',\n",
       "      'PE(P-16:0/22:4)',\n",
       "      'PE(P-16:0/22:5) (n3)',\n",
       "      'PE(P-16:0/22:5) (n6)',\n",
       "      'PE(P-16:0/22:6)',\n",
       "      'PE(P-17:0/20:4)(a)',\n",
       "      'PE(P-17:0/20:4)(b)',\n",
       "      'PE(P-17:0/22:6)(a)',\n",
       "      'PE(P-17:0/22:6)(b)',\n",
       "      'PE(P-18:0/18:1)',\n",
       "      'PE(P-18:0/18:2)',\n",
       "      'PE(P-18:0/18:3)',\n",
       "      'PE(P-18:0/20:3)(a)',\n",
       "      'PE(P-18:0/20:3)(b)',\n",
       "      'PE(P-18:0/20:4)',\n",
       "      'PE(P-18:0/20:5)',\n",
       "      'PE(P-18:0/22:4)',\n",
       "      'PE(P-18:0/22:5) (n3)',\n",
       "      'PE(P-18:0/22:5) (n6)',\n",
       "      'PE(P-18:0/22:6)',\n",
       "      'PE(P-18:1/18:1)',\n",
       "      'PE(P-18:1/18:1)(a)',\n",
       "      'PE(P-18:1/18:1)(b)',\n",
       "      'PE(P-18:1/18:2)(a)',\n",
       "      'PE(P-18:1/18:2)(b)',\n",
       "      'PE(P-18:1/18:3)',\n",
       "      'PE(P-18:1/20:3)(a)',\n",
       "      'PE(P-18:1/20:3)(b)',\n",
       "      'PE(P-18:1/20:4)(a)',\n",
       "      'PE(P-18:1/20:4)(b)',\n",
       "      'PE(P-18:1/20:5)(a)',\n",
       "      'PE(P-18:1/20:5)(b)',\n",
       "      'PE(P-18:1/22:4)',\n",
       "      'PE(P-18:1/22:5)(a)',\n",
       "      'PE(P-18:1/22:5)(b)',\n",
       "      'PE(P-18:1/22:6)(a)',\n",
       "      'PE(P-18:1/22:6)(b)',\n",
       "      'PE(P-19:0/20:4)(a)',\n",
       "      'PE(P-19:0/20:4)(b)',\n",
       "      'PE(P-20:0/18:1)',\n",
       "      'PE(P-20:0/18:2)',\n",
       "      'PE(P-20:0/20:4)',\n",
       "      'PE(P-20:0/22:6)',\n",
       "      'PE(P-20:1/20:4)',\n",
       "      'PE(P-20:1/22:6)',\n",
       "      'LPE(16:0) [sn1]',\n",
       "      'LPE(16:0) [sn2]',\n",
       "      'LPE(17:0) [sn1]',\n",
       "      'LPE(17:0) [sn2]',\n",
       "      'LPE(18:0) [sn1]',\n",
       "      'LPE(18:0) [sn2]',\n",
       "      'LPE(18:1) [sn1]',\n",
       "      'LPE(18:1) [sn2]',\n",
       "      'LPE(18:2) [sn1]',\n",
       "      'LPE(18:2) [sn2]',\n",
       "      'LPE(20:4) [sn1]',\n",
       "      'LPE(20:4) [sn2]',\n",
       "      'LPE(22:6) [sn1]',\n",
       "      'LPE(22:6) [sn2]',\n",
       "      'LPE(P-16:0)',\n",
       "      'LPE(P-18:0)',\n",
       "      'LPE(P-18:1)',\n",
       "      'LPE(P-20:0)',\n",
       "      'PI (38:5) (b)',\n",
       "      'PI(15-MHDA_18:1)/PI(17:0_18:1)',\n",
       "      'PI(15-MHDA_18:2)/PI(17:0_18:2)',\n",
       "      'PI(15-MHDA_20:4)/PI(17:0_20:4)',\n",
       "      'PI(16:0/16:0)',\n",
       "      'PI(16:0_16:1)',\n",
       "      'PI(16:0_20:3) (a)',\n",
       "      'PI(16:0_20:3) (b)',\n",
       "      'PI(16:0_20:4)',\n",
       "      'PI(18:0_18:1)',\n",
       "      'PI(18:0_20:2)',\n",
       "      'PI(18:0_20:3) (a)',\n",
       "      'PI(18:0_20:3) (b)',\n",
       "      'PI(18:0_20:4)',\n",
       "      'PI(18:0_22:4)',\n",
       "      'PI(18:0_22:5) (n3)',\n",
       "      'PI(18:0_22:5) (n6)',\n",
       "      'PI(18:0_22:6)',\n",
       "      'PI(18:1_18:2)',\n",
       "      'PI(20:0_20:4)',\n",
       "      'PI(34:0)',\n",
       "      'PI(34:1)',\n",
       "      'PI(36:2)',\n",
       "      'PI(37:6)',\n",
       "      'PI(38:5) (a)',\n",
       "      'PI(38:6)',\n",
       "      'PI(39:6)',\n",
       "      'LPI(18:0) [sn1]',\n",
       "      'LPI(18:0) [sn2]',\n",
       "      'LPI(18:1) [sn1]',\n",
       "      'LPI(18:1) [sn2]',\n",
       "      'LPI(18:2) [sn1]',\n",
       "      'LPI(18:2) [sn2]',\n",
       "      'LPI(20:4) [sn1]',\n",
       "      'LPI(20:4) [sn2]',\n",
       "      'PS(36:1)',\n",
       "      'PS(36:2)',\n",
       "      'PS(38:3)',\n",
       "      'PS(38:4)',\n",
       "      'PS(38:5)',\n",
       "      'PS(40:5)',\n",
       "      'PS(40:6)',\n",
       "      'PG(34:1)',\n",
       "      'PG(34:2)',\n",
       "      'PG(36:1)',\n",
       "      'PG(36:2)',\n",
       "      'CE(14:0)',\n",
       "      'CE(15:0)',\n",
       "      'CE(16:0)',\n",
       "      'CE(16:1)',\n",
       "      'CE(16:2)',\n",
       "      'CE(17:0)',\n",
       "      'CE(17:1)',\n",
       "      'CE(18:0)',\n",
       "      'CE(18:1)',\n",
       "      'CE(18:2)',\n",
       "      'CE(18:3)',\n",
       "      'CE(20:0)',\n",
       "      'CE(20:1)',\n",
       "      'CE(20:2)',\n",
       "      'CE(20:3)',\n",
       "      'CE(20:4)',\n",
       "      'CE(20:5)',\n",
       "      'CE(22:0)',\n",
       "      'CE(22:1)',\n",
       "      'CE(22:4)',\n",
       "      'CE(22:5) (n3)',\n",
       "      'CE(22:5) (n6)',\n",
       "      'CE(22:6)',\n",
       "      'CE(24:0)',\n",
       "      'CE(24:1)',\n",
       "      'CE(24:4)',\n",
       "      'CE(24:5)',\n",
       "      'CE(24:6)',\n",
       "      'COH',\n",
       "      'DE(16:0)',\n",
       "      'DE(18:1)',\n",
       "      'DE(18:2)',\n",
       "      'DE(20:4)',\n",
       "      'DE(20:5)',\n",
       "      'DE(22:6)',\n",
       "      'Acylcarnitine(12:0)',\n",
       "      'Acylcarnitine(13:0)',\n",
       "      'Acylcarnitine(14:0)',\n",
       "      'Acylcarnitine(14:1)',\n",
       "      'Acylcarnitine(14:2)',\n",
       "      'Acylcarnitine(15:0) (a)',\n",
       "      'Acylcarnitine(15:0) (b)',\n",
       "      'Acylcarnitine(16:0)',\n",
       "      'Acylcarnitine(16:1)',\n",
       "      'Acylcarnitine(17:0) (a)',\n",
       "      'Acylcarnitine(17:0) (b)',\n",
       "      'Acylcarnitine(18:0)',\n",
       "      'Acylcarnitine(18:1)',\n",
       "      'Acylcarnitine(18:2)',\n",
       "      'DG(14:0_16:0)',\n",
       "      'DG(14:0_18:2)',\n",
       "      'DG(16:0_16:0)',\n",
       "      'DG(16:0_16:1)',\n",
       "      'DG(16:0_18:1)',\n",
       "      'DG(16:0_18:2)',\n",
       "      'DG(16:0_20:4)',\n",
       "      'DG(16:0_22:5)',\n",
       "      'DG(16:0_22:6)',\n",
       "      'DG(16:1_18:1)',\n",
       "      'DG(18:0_18:1)',\n",
       "      'DG(18:0_18:2)',\n",
       "      'DG(18:0_20:4)',\n",
       "      'DG(18:0_22:6)',\n",
       "      'DG(18:1_18:1)',\n",
       "      'DG(18:1_18:2)',\n",
       "      'DG(18:1_18:3)',\n",
       "      'DG(18:1_20:3)',\n",
       "      'DG(18:1_20:4)',\n",
       "      'DG(18:1_20:5)',\n",
       "      'DG(18:1_22:5)',\n",
       "      'DG(18:1_22:6)',\n",
       "      'DG(18:2_18:2)',\n",
       "      'DG(18:2_20:4)',\n",
       "      'DG(18:2_22:6)',\n",
       "      'TG(48:0) [NL-16:0]',\n",
       "      'TG(48:0) [NL-18:0]',\n",
       "      'TG(48:0) [SIM]',\n",
       "      'TG(48:1) [NL-16:1]',\n",
       "      'TG(48:1) [NL-18:1]',\n",
       "      'TG(48:1) [SIM]',\n",
       "      'TG(48:2) [NL-14:0]',\n",
       "      'TG(48:2) [NL-14:1]',\n",
       "      'TG(48:2) [NL-16:1]',\n",
       "      'TG(48:2) [NL-18:2]',\n",
       "      'TG(48:2) [SIM]',\n",
       "      'TG(48:3) [NL-14:0]',\n",
       "      'TG(48:3) [NL-16:1]',\n",
       "      'TG(48:3) [NL-18:3]',\n",
       "      'TG(48:3) [SIM]',\n",
       "      'TG(49:1) [NL-16:1]',\n",
       "      'TG(49:1) [NL-17:1]',\n",
       "      'TG(49:1) [SIM]',\n",
       "      'TG(50:0) [NL-18:0]',\n",
       "      'TG(50:0) [SIM]',\n",
       "      'TG(50:1) [NL-14:0]',\n",
       "      'TG(50:1) [NL-16:0]',\n",
       "      'TG(50:1) [NL-18:1]',\n",
       "      'TG(50:1) [SIM]',\n",
       "      'TG(50:2) [NL-14:0]',\n",
       "      'TG(50:2) [NL-16:1]',\n",
       "      'TG(50:2) [NL-18:1]',\n",
       "      'TG(50:2) [NL-18:2]',\n",
       "      'TG(50:2) [SIM]',\n",
       "      'TG(50:3) [NL-14:0]',\n",
       "      'TG(50:3) [NL-14:1]',\n",
       "      'TG(50:3) [NL-16:1]',\n",
       "      'TG(50:3) [NL-18:2]',\n",
       "      'TG(50:3) [NL-18:3]',\n",
       "      'TG(50:3) [SIM]',\n",
       "      'TG(50:4) [NL-14:0]',\n",
       "      'TG(50:4) [NL-18:3]',\n",
       "      'TG(50:4) [NL-20:4]',\n",
       "      'TG(50:4) [SIM]',\n",
       "      'TG(51:0) [NL-16:0]',\n",
       "      'TG(51:1) [NL-17:0]',\n",
       "      'TG(51:1) [SIM]',\n",
       "      'TG(51:2) [NL-15:0]',\n",
       "      'TG(51:2) [NL-17:0]',\n",
       "      'TG(51:2) [NL-17:1]',\n",
       "      'TG(51:2) [SIM]',\n",
       "      'TG(52:1) [NL-18:0]',\n",
       "      'TG(52:1) [NL-18:1]',\n",
       "      'TG(52:1) [SIM]',\n",
       "      'TG(52:2) [NL-16:0]',\n",
       "      'TG(52:2) [NL-18:2]',\n",
       "      'TG(52:2) [SIM]',\n",
       "      'TG(52:3) [NL-16:1]',\n",
       "      'TG(52:3) [NL-18:2]',\n",
       "      'TG(52:3) [SIM]',\n",
       "      'TG(52:4) [NL-16:1]',\n",
       "      'TG(52:4) [NL-18:2]',\n",
       "      'TG(52:4) [NL-18:3]',\n",
       "      'TG(52:4) [SIM]',\n",
       "      'TG(52:5) [NL-18:3]',\n",
       "      'TG(52:5) [NL-20:4]',\n",
       "      'TG(52:5) [NL-20:5]',\n",
       "      'TG(52:5) [SIM]',\n",
       "      'TG(53:2) [NL-17:1]',\n",
       "      'TG(53:2) [NL-18:1]',\n",
       "      'TG(53:2) [SIM]',\n",
       "      'TG(54:0) [NL-18:0]',\n",
       "      'TG(54:0) [SIM]',\n",
       "      'TG(54:1) [NL-18:1]',\n",
       "      'TG(54:1) [SIM]',\n",
       "      'TG(54:2) [NL-18:0]',\n",
       "      'TG(54:2) [NL-20:1]',\n",
       "      'TG(54:2) [SIM]',\n",
       "      'TG(54:3) [NL-18:1]',\n",
       "      'TG(54:3) [NL-18:2]',\n",
       "      'TG(54:3) [SIM]',\n",
       "      'TG(54:4) [NL-18:2]',\n",
       "      'TG(54:4) [NL-20:3]',\n",
       "      'TG(54:4) [SIM]',\n",
       "      'TG(54:5) [NL-18:3]',\n",
       "      'TG(54:5) [NL-20:4]',\n",
       "      'TG(54:5) [SIM]',\n",
       "      'TG(54:6) [NL-18:3]',\n",
       "      'TG(54:6) [NL-20:4]',\n",
       "      'TG(54:6) [NL-20:5]',\n",
       "      'TG(54:6) [NL-22:6]',\n",
       "      'TG(54:6) [SIM]',\n",
       "      'TG(54:7) [NL-20:5]',\n",
       "      'TG(54:7) [NL-22:6]',\n",
       "      'TG(54:7) [SIM]',\n",
       "      'TG(56:6) [NL-20:4]',\n",
       "      'TG(56:6) [NL-22:5](a)',\n",
       "      'TG(56:6) [NL-22:5](b)',\n",
       "      'TG(56:6) [SIM]',\n",
       "      'TG(56:7) [NL-20:4]',\n",
       "      'TG(56:7) [NL-20:5]',\n",
       "      'TG(56:7) [NL-22:5](a)',\n",
       "      'TG(56:7) [NL-22:5](b)',\n",
       "      'TG(56:7) [NL-22:6]',\n",
       "      'TG(56:7) [SIM]',\n",
       "      'TG(56:8) [NL-20:4]',\n",
       "      'TG(56:8) [NL-20:5]',\n",
       "      'TG(56:8) [NL-22:6]',\n",
       "      'TG(56:8) [SIM]',\n",
       "      'TG(56:9) [NL-22:6]',\n",
       "      'TG(56:9) [SIM]',\n",
       "      'TG(58:10) [NL-22:6]',\n",
       "      'TG(58:10) [SIM]',\n",
       "      'TG(58:8) [NL-22:6]',\n",
       "      'TG(58:8) [SIM]',\n",
       "      'TG(58:9) [NL-22:6]',\n",
       "      'TG(58:9) [SIM]',\n",
       "      'TG(O-50:1) [NL-15:0]',\n",
       "      'TG(O-50:1) [NL-16:0]',\n",
       "      'TG(O-50:1) [NL-17:1]',\n",
       "      'TG(O-50:1) [NL-18:1]',\n",
       "      'TG(O-50:1) [SIM]',\n",
       "      'TG(O-50:2) [NL-16:1]',\n",
       "      'TG(O-50:2) [NL-18:1]',\n",
       "      'TG(O-50:2) [NL-18:2]',\n",
       "      'TG(O-50:2) [SIM]',\n",
       "      'TG(O-50:3) [NL-18:2]',\n",
       "      'TG(O-50:3) [SIM]',\n",
       "      'TG(O-52:0) [NL-16:0]',\n",
       "      'TG(O-52:1) [NL-16:0]',\n",
       "      'TG(O-52:1) [NL-18:1]',\n",
       "      'TG(O-52:1) [SIM]',\n",
       "      'TG(O-52:2) [NL-16:0]',\n",
       "      'TG(O-52:2) [NL-17:1]',\n",
       "      'TG(O-52:2) [NL-18:1]',\n",
       "      'TG(O-52:2) [SIM]',\n",
       "      'TG(O-54:2) [NL-17:1]',\n",
       "      'TG(O-54:2) [NL-18:1]',\n",
       "      'TG(O-54:2) [SIM]',\n",
       "      'TG(O-54:3) [NL-17:1]',\n",
       "      'TG(O-54:3) [NL-18:1]',\n",
       "      'TG(O-54:3) [SIM]',\n",
       "      'TG(O-54:4) [NL-17:1]',\n",
       "      'TG(O-54:4) [NL-18:2]',\n",
       "      'TG(O-54:4) [SIM]',\n",
       "      'Ubiquinone']},\n",
       "    'type': 'array'},\n",
       "   'samples': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type', 'submitter_id', 'assay_description'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Lipidomics Assay',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'metabolomics_file.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'data_file',\n",
       "  'description': 'Data file that contains the results of a metabolomics assay',\n",
       "  'id': 'metabolomics_file',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'metabolomics_files',\n",
       "      'label': 'derives_from',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'metabolomics_assays',\n",
       "      'required': False,\n",
       "      'target_type': 'metabolomics_assay'},\n",
       "     {'backref': 'metabolomics_files',\n",
       "      'label': 'describes',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'core_metadata_collections',\n",
       "      'required': False,\n",
       "      'target_type': 'core_metadata_collection'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/data_file_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'core_metadata_collections': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'cv': {'description': 'Coefficient of variation (%)',\n",
       "    'type': ['number', 'null']},\n",
       "   'data_category': {'description': 'Broad categorization of the contents of the data file.',\n",
       "    'enum': ['mass spec raw', 'mass spec analysed', 'summarised results']},\n",
       "   'data_format': {'description': 'The format of the data in this data file',\n",
       "    'enum': ['wiff']},\n",
       "   'data_type': {'description': 'The type of data in this data file',\n",
       "    'enum': ['MS', 'MS/MS']},\n",
       "   'metabolomic_unit': {'description': 'Units of measurement captured in the metabolomics file.',\n",
       "    'enum': ['pmol/mL', 'umol/mL']},\n",
       "   'metabolomics_assays': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'data_category',\n",
       "   'data_format',\n",
       "   'data_type'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Metabolomics File',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'acknowledgement.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'administrative',\n",
       "  'description': 'Acknowledgement of an individual or group involved in a project.',\n",
       "  'id': 'acknowledgement',\n",
       "  'links': [{'backref': 'acknowledgements',\n",
       "    'label': 'contribute_to',\n",
       "    'multiplicity': 'many_to_many',\n",
       "    'name': 'projects',\n",
       "    'required': True,\n",
       "    'target_type': 'project'}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'acknowledgee': {'description': 'Name of the individual or group to be acknowledged. Preferably as First name Last name',\n",
       "    'type': 'string'},\n",
       "   'contact_type': {'description': 'The type of contact or role in the project, e.g. Principal Investigator',\n",
       "    'enum': ['Principal Investigator', 'Data custodian', 'Co-investigator']},\n",
       "   'orcid': {'description': 'The ORCID number for the acknowledgee',\n",
       "    'pattern': '^[0]{4}-\\\\d{4}-\\\\d{4}-(\\\\d{3}X|\\\\d{4})$',\n",
       "    'type': 'string'},\n",
       "   'projects': {'$ref': '_definitions.yaml#/to_many_project'}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'projects',\n",
       "   'acknowledgee',\n",
       "   'contact_type'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Acknowledgement',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'medical_history.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'clinical',\n",
       "  'description': 'Medical history of the participant',\n",
       "  'id': 'medical_history',\n",
       "  'links': [{'backref': 'medical_histories',\n",
       "    'label': 'describes',\n",
       "    'multiplicity': 'one_to_one',\n",
       "    'name': 'subjects',\n",
       "    'required': True,\n",
       "    'target_type': 'subject'}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'atrial_fibrillation': {'description': 'Self reported Atrial Fibrillation',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'cabg': {'description': 'Has the patient had Coronary Artery Bypass Graft (CABG) Surgery?',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'cvd_death': {'description': 'Death due to CVD ICD10: I10-I99',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'cvd_family_history': {'description': 'Does the participant have a family history of CVD? (Family history of premature CVD is defined as coronary heart disease or stroke in a first-degree female relative aged <65 years or a firstdegree male relative aged <55 years.)',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'cvd_non_fatal_10_year': {'description': 'Non fatal cardiovascular disease at 10 years.',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'cvd_reported': {'description': 'Reported Stroke/MI/Angina at baseline, by clinician',\n",
       "    'enum': ['yes', 'no'],\n",
       "    'termDef': [{'source': 'hpo', 'term': 'Stroke', 'term_id': 'HP:0001297'}]},\n",
       "   'cvd_self_reported': {'description': 'Self-reported Stroke/MI/Angina at baseline by the participant.',\n",
       "    'enum': ['yes', 'no'],\n",
       "    'termDef': [{'source': 'hpo', 'term': 'Stroke', 'term_id': 'HP:0001297'}]},\n",
       "   'date_death': {'description': 'Date of death in the format YYYY-MM-DD',\n",
       "    'pattern': '^\\\\d{4}-\\\\d{2}-\\\\d{2}',\n",
       "    'type': 'string'},\n",
       "   'diabetes_self_reported': {'description': 'Self-reported diabetes in the participant',\n",
       "    'enum': ['yes', 'no'],\n",
       "    'termDef': [{'source': 'hpo',\n",
       "      'term': 'Diabetes mellitus',\n",
       "      'term_id': 'HP:0000819'}]},\n",
       "   'diabetes_type': {'description': 'Diabetes mellitus type.',\n",
       "    'enum': ['type-1', 'type-2'],\n",
       "    'termDef': [{'source': 'hpo',\n",
       "      'term': 'Diabetes mellitus',\n",
       "      'term_id': 'HP:0000819',\n",
       "      'term_version': '2021-10-10'}]},\n",
       "   'heart_failure_status': {'description': 'Self reported heart failure status.',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'heart_rate': {'description': 'Indicate the number of heart beats over a period of one minute.',\n",
       "    'type': 'integer'},\n",
       "   'hypertension_measured': {'description': 'Hypertension reported by a clinician at recruitment. Systolic BP >140mmHg, diastolic BP >90mmHg, or both',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'hypertension_self_reported': {'description': 'Hypertension self reported by patient. Systolic BP >140mmHg, diastolic BP >90mmHg, or both',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'mhx_pad': {'description': 'Vascular history - Code 13. Peripheral vascular disease (includes abdominal aortic aneurism)',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'os_event': {'description': 'Overall survival or vital status ',\n",
       "    'enum': ['alive', 'dead', 'censored']},\n",
       "   'os_time': {'description': 'Overall survival time in months. Calculated as the time between baseline date and death date.',\n",
       "    'type': 'integer'},\n",
       "   'stent': {'description': 'Coronary stent', 'enum': ['yes', 'no']},\n",
       "   'subjects': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Medical History',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " '_definitions.yaml': {'UUID': {'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "   'term': {'$ref': '_terms.yaml#/UUID'},\n",
       "   'type': 'string'},\n",
       "  'data_bundle_state': {'default': 'submitted',\n",
       "   'description': 'State of a data bundle.',\n",
       "   'enum': ['submitted',\n",
       "    'validated',\n",
       "    'error',\n",
       "    'released',\n",
       "    'suppressed',\n",
       "    'redacted']},\n",
       "  'data_file_error_type': {'enum': ['file_size', 'file_format', 'md5sum'],\n",
       "   'term': {'$ref': '_terms.yaml#/data_file_error_type'}},\n",
       "  'data_file_properties': {'$ref': '#/ubiquitous_properties',\n",
       "   'error_type': {'$ref': '#/data_file_error_type'},\n",
       "   'file_format': {'$ref': '#/file_format'},\n",
       "   'file_name': {'$ref': '#/file_name'},\n",
       "   'file_size': {'$ref': '#/file_size'},\n",
       "   'file_state': {'$ref': '#/file_state'},\n",
       "   'ga4gh_drs_uri': {'$ref': '#/ga4gh_drs_uri'},\n",
       "   'md5sum': {'$ref': '#/md5sum'},\n",
       "   'object_id': {'$ref': '#/object_id'}},\n",
       "  'date': {'description': 'Date in YYYY-MM-DD format.',\n",
       "   'oneOf': [{'format': 'date', 'type': 'string'}, {'type': 'null'}],\n",
       "   'pattern': '^\\\\d{4}-\\\\d{2}-\\\\d{2}$'},\n",
       "  'datetime': {'oneOf': [{'format': 'date-time', 'type': 'string'},\n",
       "    {'type': 'null'}],\n",
       "   'term': {'$ref': '_terms.yaml#/datetime'}},\n",
       "  'file_format': {'term': {'$ref': '_terms.yaml#/file_format'},\n",
       "   'type': 'string'},\n",
       "  'file_name': {'term': {'$ref': '_terms.yaml#/file_name'}, 'type': 'string'},\n",
       "  'file_size': {'term': {'$ref': '_terms.yaml#/file_size'}, 'type': 'integer'},\n",
       "  'file_state': {'default': 'registered',\n",
       "   'enum': ['registered',\n",
       "    'uploading',\n",
       "    'uploaded',\n",
       "    'validating',\n",
       "    'validated',\n",
       "    'submitted',\n",
       "    'processing',\n",
       "    'processed',\n",
       "    'released',\n",
       "    'error'],\n",
       "   'term': {'$ref': '_terms.yaml#/file_state'}},\n",
       "  'foreign_key': {'additionalProperties': True,\n",
       "   'properties': {'id': {'$ref': '#/UUID'},\n",
       "    'submitter_id': {'type': 'string'}},\n",
       "   'type': 'object'},\n",
       "  'foreign_key_project': {'additionalProperties': True,\n",
       "   'properties': {'code': {'type': 'string'}, 'id': {'$ref': '#/UUID'}},\n",
       "   'type': 'object'},\n",
       "  'ga4gh_drs_uri': {'term': {'$ref': '_terms.yaml#/ga4gh_drs_uri'},\n",
       "   'type': 'string'},\n",
       "  'id': '_definitions',\n",
       "  'md5sum': {'pattern': '^[a-f0-9]{32}$',\n",
       "   'term': {'$ref': '_terms.yaml#/md5sum'},\n",
       "   'type': 'string'},\n",
       "  'month': {'description': 'month (MM)',\n",
       "   'pattern': '^\\\\d{2}$',\n",
       "   'type': 'string'},\n",
       "  'object_id': {'description': 'The GUID of the object in the index service.',\n",
       "   'type': 'string'},\n",
       "  'parent_uuids': {'items': {'$ref': '#/UUID'},\n",
       "   'minItems': 1,\n",
       "   'type': 'array',\n",
       "   'uniqueItems': True},\n",
       "  'postcode': {'description': 'Post code (DDDD)',\n",
       "   'pattern': '^\\\\d{4}$',\n",
       "   'type': 'string'},\n",
       "  'project_id': {'term': {'$ref': '_terms.yaml#/project_id'},\n",
       "   'type': 'string'},\n",
       "  'qc_metrics_state': {'enum': ['FAIL', 'PASS', 'WARN'],\n",
       "   'term': {'$ref': '_terms.yaml#/qc_metric_state'}},\n",
       "  'release_state': {'default': 'unreleased',\n",
       "   'description': 'Release state of an entity.',\n",
       "   'enum': ['unreleased', 'released', 'redacted']},\n",
       "  'state': {'default': 'validated',\n",
       "   'downloadable': ['uploaded',\n",
       "    'md5summed',\n",
       "    'validating',\n",
       "    'validated',\n",
       "    'error',\n",
       "    'invalid',\n",
       "    'released'],\n",
       "   'oneOf': [{'enum': ['uploading',\n",
       "      'uploaded',\n",
       "      'md5summing',\n",
       "      'md5summed',\n",
       "      'validating',\n",
       "      'error',\n",
       "      'invalid',\n",
       "      'suppressed',\n",
       "      'redacted',\n",
       "      'live']},\n",
       "    {'enum': ['validated', 'submitted', 'released']}],\n",
       "   'public': ['live'],\n",
       "   'term': {'$ref': '_terms.yaml#/state'}},\n",
       "  'to_many': {'anyOf': [{'items': {'$ref': '#/foreign_key', 'minItems': 1},\n",
       "     'type': 'array'},\n",
       "    {'$ref': '#/foreign_key'}]},\n",
       "  'to_many_project': {'anyOf': [{'items': {'$ref': '#/foreign_key_project',\n",
       "      'minItems': 1},\n",
       "     'type': 'array'},\n",
       "    {'$ref': '#/foreign_key_project'}]},\n",
       "  'to_one': {'anyOf': [{'items': {'$ref': '#/foreign_key',\n",
       "      'maxItems': 1,\n",
       "      'minItems': 1},\n",
       "     'type': 'array'},\n",
       "    {'$ref': '#/foreign_key'}]},\n",
       "  'to_one_project': {'anyOf': [{'items': {'$ref': '#/foreign_key_project',\n",
       "      'maxItems': 1,\n",
       "      'minItems': 1},\n",
       "     'type': 'array'},\n",
       "    {'$ref': '#/foreign_key_project'}]},\n",
       "  'ubiquitous_properties': {'created_datetime': {'$ref': '#/datetime'},\n",
       "   'id': {'$ref': '#/UUID', 'systemAlias': 'node_id'},\n",
       "   'project_id': {'$ref': '#/project_id'},\n",
       "   'state': {'$ref': '#/state'},\n",
       "   'submitter_id': {'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n',\n",
       "    'type': ['string']},\n",
       "   'type': {'type': 'string'},\n",
       "   'updated_datetime': {'$ref': '#/datetime'}},\n",
       "  'workflow_properties': {'$ref': '#/ubiquitous_properties',\n",
       "   'workflow_end_datetime': {'$ref': '#/datetime'},\n",
       "   'workflow_link': {'description': 'Link to Github hash for the CWL workflow used.',\n",
       "    'type': 'string'},\n",
       "   'workflow_start_datetime': {'$ref': '#/datetime'},\n",
       "   'workflow_version': {'description': 'Major version for a GDC workflow.',\n",
       "    'type': 'string'}},\n",
       "  'year': {'description': 'Year (YYYY)',\n",
       "   'pattern': '^\\\\d{4}$',\n",
       "   'type': 'string'}},\n",
       " '_settings.yaml': {'enable_case_cache': False,\n",
       "  '_dict_commit': '520a25999fd183f6c5b7ddef2980f3e839517da5',\n",
       "  '_dict_version': '0.2.1-9-g520a259'},\n",
       " 'blood_pressure_test.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'clinical',\n",
       "  'description': 'Blood pressure reading (insert method here).',\n",
       "  'id': 'blood_pressure_test',\n",
       "  'links': [{'backref': 'blood_pressure_tests',\n",
       "    'label': 'taken_by',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'name': 'subjects',\n",
       "    'required': True,\n",
       "    'target_type': 'subject'}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'bp_diastolic': {'description': 'Resting diastolic blood pressure from the upper arm in a clinical setting. (mmHg)',\n",
       "    'termDef': [{'source': 'efo',\n",
       "      'term': 'diastolic blood pressure',\n",
       "      'term_id': 'EFO:0006336'}],\n",
       "    'type': 'integer'},\n",
       "   'bp_systolic': {'description': 'Resting systolic blood pressure from the upper arm in a clinical setting. (mmHg)',\n",
       "    'termDef': [{'source': 'efo',\n",
       "      'term': 'systolic blood pressure',\n",
       "      'term_id': 'EFO:0006335'}],\n",
       "    'type': 'integer'},\n",
       "   'subjects': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Blood Pressure Test',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'genomics_assay.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'experimental_methods',\n",
       "  'description': 'Details about the methods used to produce genomic output files',\n",
       "  'id': 'genomics_assay',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'genomics_assays',\n",
       "      'label': 'generated_from',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'samples',\n",
       "      'required': False,\n",
       "      'target_type': 'sample'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'instrument_model': {'description': 'The name and/or number associated with a specific sequencing instrument model',\n",
       "    'enum': ['454 GS',\n",
       "     '454 GS 20',\n",
       "     '454 GS FLX',\n",
       "     '454 GS FLX Titanium',\n",
       "     '454 GS FLX+',\n",
       "     '454 GS Junior',\n",
       "     'AB 310 Genetic Analyzer',\n",
       "     'AB 3130 Genetic Analyzer',\n",
       "     'AB 3130xL Genetic Analyzer',\n",
       "     'AB 3500 Genetic Analyzer',\n",
       "     'AB 3500xL Genetic Analyzer',\n",
       "     'AB 3730 Genetic Analyzer',\n",
       "     'AB 3730xL Genetic Analyzer',\n",
       "     'AB 5500 Genetic Analyzer',\n",
       "     'AB 5500xl Genetic Analyzer',\n",
       "     'AB 5500xl-W Genetic Analysis System',\n",
       "     'AB SOLiD 3 Plus System',\n",
       "     'AB SOLiD 4 System',\n",
       "     'AB SOLiD 4hq System',\n",
       "     'AB SOLiD PI System',\n",
       "     'AB SOLiD System',\n",
       "     'AB SOLiD System 2.0',\n",
       "     'AB SOLiD System 3.0',\n",
       "     'BGISEQ-50',\n",
       "     'BGISEQ-500',\n",
       "     'Complete Genomics',\n",
       "     'DNBSEQ-G400',\n",
       "     'DNBSEQ-G400 FAST',\n",
       "     'DNBSEQ-G50',\n",
       "     'DNBSEQ-T7',\n",
       "     'Element AVITI',\n",
       "     'FASTASeq 300',\n",
       "     'GENIUS',\n",
       "     'GS111',\n",
       "     'Genapsys Sequencer',\n",
       "     'GenoCare 1600',\n",
       "     'GenoLab M',\n",
       "     'GridION',\n",
       "     'Helicos HeliScope',\n",
       "     'HiSeq X Five',\n",
       "     'HiSeq X Ten',\n",
       "     'Illumina Genome Analyzer',\n",
       "     'Illumina Genome Analyzer II',\n",
       "     'Illumina Genome Analyzer IIx',\n",
       "     'Illumina HiScanSQ',\n",
       "     'Illumina HiSeq 1000',\n",
       "     'Illumina HiSeq 1500',\n",
       "     'Illumina HiSeq 2000',\n",
       "     'Illumina HiSeq 2500',\n",
       "     'Illumina HiSeq 3000',\n",
       "     'Illumina HiSeq 4000',\n",
       "     'Illumina HiSeq X',\n",
       "     'Illumina MiSeq',\n",
       "     'Illumina MiniSeq',\n",
       "     'Illumina NovaSeq 6000',\n",
       "     'Illumina NovaSeq X',\n",
       "     'Illumina iSeq 100',\n",
       "     'Ion GeneStudio S5',\n",
       "     'Ion GeneStudio S5 Plus',\n",
       "     'Ion GeneStudio S5 Prime',\n",
       "     'Ion Torrent Genexus',\n",
       "     'Ion Torrent PGM',\n",
       "     'Ion Torrent Proton',\n",
       "     'Ion Torrent S5',\n",
       "     'Ion Torrent S5 XL',\n",
       "     'MGISEQ-2000RS',\n",
       "     'MinION',\n",
       "     'NextSeq 1000',\n",
       "     'NextSeq 2000',\n",
       "     'NextSeq 500',\n",
       "     'NextSeq 550',\n",
       "     'Onso',\n",
       "     'PacBio RS',\n",
       "     'PacBio RS II',\n",
       "     'PromethION',\n",
       "     'Revio',\n",
       "     'Sentosa SQ301',\n",
       "     'Sequel',\n",
       "     'Sequel II',\n",
       "     'Sequel IIe',\n",
       "     'Tapestri',\n",
       "     'UG 100',\n",
       "     'unspecified'],\n",
       "    'enumDef': [{'enumeration': '454 GS',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': '454 GS 20',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': '454 GS FLX',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': '454 GS FLX Titanium',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': '454 GS FLX+',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': '454 GS Junior',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB 310 Genetic Analyzer',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB 3130 Genetic Analyzer',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB 3130xL Genetic Analyzer',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB 3500 Genetic Analyzer',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB 3500xL Genetic Analyzer',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB 3730 Genetic Analyzer',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB 3730xL Genetic Analyzer',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB 5500 Genetic Analyzer',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB 5500xl Genetic Analyzer',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB 5500xl-W Genetic Analysis System',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB SOLiD 3 Plus System',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB SOLiD 4 System',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB SOLiD 4hq System',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB SOLiD PI System',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB SOLiD System',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB SOLiD System 2.0',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'AB SOLiD System 3.0',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'BGISEQ-50',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'BGISEQ-500',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Complete Genomics',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'DNBSEQ-G400',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'DNBSEQ-G400 FAST',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'DNBSEQ-G50',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'DNBSEQ-T7',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Element AVITI',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'FASTASeq 300',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'GENIUS',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'GS111',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Genapsys Sequencer',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'GenoCare 1600',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'GenoLab M',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'GridION',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Helicos HeliScope',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'HiSeq X Five',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'HiSeq X Ten',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina Genome Analyzer',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina Genome Analyzer II',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina Genome Analyzer IIx',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina HiScanSQ',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina HiSeq 1000',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina HiSeq 1500',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina HiSeq 2000',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina HiSeq 2500',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina HiSeq 3000',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina HiSeq 4000',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina HiSeq X',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina MiSeq',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina MiniSeq',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina NovaSeq 6000',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina NovaSeq X',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Illumina iSeq 100',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Ion GeneStudio S5',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Ion GeneStudio S5 Plus',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Ion GeneStudio S5 Prime',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Ion Torrent Genexus',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Ion Torrent PGM',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Ion Torrent Proton',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Ion Torrent S5',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Ion Torrent S5 XL',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'MGISEQ-2000RS',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'MinION',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'NextSeq 1000',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'NextSeq 2000',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'NextSeq 500',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'NextSeq 550',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Onso',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'PacBio RS',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'PacBio RS II',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'PromethION',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Revio',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Sentosa SQ301',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Sequel',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Sequel II',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Sequel IIe',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'Tapestri',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'UG 100',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'unspecified',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_model',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'}]},\n",
       "   'instrument_platform': {'description': 'The name of the technology platform used to perform nucleic acid sequencing or quantification',\n",
       "    'enum': ['ABI_SOLID',\n",
       "     'BGISEQ',\n",
       "     'CAPILLARY',\n",
       "     'COMPLETE_GENOMICS',\n",
       "     'DNBSEQ',\n",
       "     'ELEMENT',\n",
       "     'GENAPSYS',\n",
       "     'GENE_MIND',\n",
       "     'HELICOS',\n",
       "     'ILLUMINA',\n",
       "     'ION_TORRENT',\n",
       "     'LS454',\n",
       "     'OXFORD_NANOPORE',\n",
       "     'PACBIO_SMRT',\n",
       "     'TAPESTRI',\n",
       "     'ULTIMA',\n",
       "     'VELA_DIAGNOSTICS'],\n",
       "    'enumDef': [{'enumeration': 'ABI_SOLID',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'BGISEQ',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'CAPILLARY',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'COMPLETE_GENOMICS',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'DNBSEQ',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'ELEMENT',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'GENAPSYS',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'GENE_MIND',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'HELICOS',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'ILLUMINA',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'ION_TORRENT',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'LS454',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'OXFORD_NANOPORE',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'PACBIO_SMRT',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'TAPESTRI',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'ULTIMA',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'},\n",
       "     {'enumeration': 'VELA_DIAGNOSTICS',\n",
       "      'source': 'https://github.com/enasequence/ena-experiment-checklist/blob/main/data/output_md/GENOMIC.md',\n",
       "      'term_id': 'instrument_platform',\n",
       "      'version': '9efee515e2a8351e767f7cfb2d7dffb138c98f82'}]},\n",
       "   'library_prep_protocol': {'description': 'Name of the protocol used to prepare the samples',\n",
       "    'type': 'string'},\n",
       "   'library_strategy': {'description': 'Sequencing technique used for this library',\n",
       "    'enum': ['WGS',\n",
       "     'WXS',\n",
       "     'LC-WGS',\n",
       "     'RNA-Seq',\n",
       "     'miRNA-Seq',\n",
       "     'Total-RNA-Seq',\n",
       "     'Targeted',\n",
       "     'Other']},\n",
       "   'samples': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'sequencing_protocol': {'description': 'Name of the sequencing protocol used to convert cleaned libraries into sequenced data',\n",
       "    'type': 'string'}},\n",
       "  'required': ['type', 'submitter_id'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Genomics Assay',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'variant_file.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'data_file',\n",
       "  'description': 'Data file that contains the results of variant calling on a SNP array or sequencing data',\n",
       "  'id': 'variant_file',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'variant_files',\n",
       "      'label': 'generated_from',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'variant_workflows',\n",
       "      'required': False,\n",
       "      'target_type': 'variant_workflow'},\n",
       "     {'backref': 'variant_files',\n",
       "      'label': 'generated_from',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'aligned_reads_files',\n",
       "      'required': False,\n",
       "      'target_type': 'aligned_reads_file'},\n",
       "     {'backref': 'variant_files',\n",
       "      'label': 'data_from',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'core_metadata_collections',\n",
       "      'required': False,\n",
       "      'target_type': 'core_metadata_collection'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/data_file_properties',\n",
       "   'aligned_reads_files': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'core_metadata_collections': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'data_category': {'description': 'Broad categorization of the contents of the data file.',\n",
       "    'enum': ['analysis',\n",
       "     'sequencing reads',\n",
       "     'single nucleotide variation',\n",
       "     'transcriptome profiling',\n",
       "     'supplemental']},\n",
       "   'data_format': {'description': 'Format of the data files.',\n",
       "    'enum': ['bed', 'vcf', 'vcf.gz']},\n",
       "   'data_type': {'description': 'Specific content type of the data file.',\n",
       "    'enum': ['unaligned reads',\n",
       "     'aligned reads',\n",
       "     'variants annotation',\n",
       "     'clinical supplement']},\n",
       "   'reference_genome_build': {'description': 'Reference genome used e.g. GRCh37.',\n",
       "    'pattern': '^GRCh[0-9]{2}$',\n",
       "    'type': 'string'},\n",
       "   'variant_workflows': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'data_category',\n",
       "   'data_format',\n",
       "   'data_type'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Variant File',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'program.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'administrative',\n",
       "  'description': 'A broad framework of goals to be achieved. (NCIt C52647)\\n',\n",
       "  'id': 'program',\n",
       "  'links': [],\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'dbgap_accession_number': {'description': 'The dbgap accession number provided for the program.',\n",
       "    'type': 'string'},\n",
       "   'id': {'$ref': '_definitions.yaml#/UUID', 'systemAlias': 'node_id'},\n",
       "   'name': {'description': 'Full name/title of the program.',\n",
       "    'type': 'string'},\n",
       "   'type': {'type': 'string'}},\n",
       "  'required': ['name', 'dbgap_accession_number'],\n",
       "  'submittable': False,\n",
       "  'systemProperties': ['id'],\n",
       "  'title': 'Program',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['name']],\n",
       "  'validators': None},\n",
       " 'serum_marker_file.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'data_file',\n",
       "  'description': 'Data file that contains the results of a serum marker assay',\n",
       "  'id': 'serum_marker_file',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'serum_marker_files',\n",
       "      'label': 'derives_from',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'serum_marker_assays',\n",
       "      'required': False,\n",
       "      'target_type': 'serum_marker_assay'},\n",
       "     {'backref': 'serum_marker_files',\n",
       "      'label': 'describes',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'core_metadata_collections',\n",
       "      'required': False,\n",
       "      'target_type': 'core_metadata_collection'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/data_file_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'core_metadata_collections': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'data_category': {'description': 'Broad categorization of the contents of the data file.',\n",
       "    'enum': ['mass spec raw', 'mass spec analysed', 'summarised results']},\n",
       "   'data_format': {'description': 'The format of the data in this data file',\n",
       "    'enum': ['csv', 'tsv', 'txt']},\n",
       "   'data_type': {'description': 'The type of data in this data file',\n",
       "    'enum': ['MS', 'MS/MS']},\n",
       "   'serum_marker_assays': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'serum_unit': {'description': 'Units of measurement captured in the file.',\n",
       "    'type': 'string'}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'data_category',\n",
       "   'data_format',\n",
       "   'data_type'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Serum Marker File',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'proteomics_assay.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'experimental_methods',\n",
       "  'description': 'Details about the methods that produced the proteomics data',\n",
       "  'id': 'proteomics_assay',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'proteomics_assays',\n",
       "      'label': 'produces',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'samples',\n",
       "      'required': False,\n",
       "      'target_type': 'sample'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'assay_description': {'description': 'A description of the type of assay used.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'protein_names': {'description': 'A list of all proteins assayed in this method',\n",
       "    'items': {'enum': ['A1BG',\n",
       "      'A2M',\n",
       "      'A2M;PZP',\n",
       "      'ADIPOQ',\n",
       "      'AFM',\n",
       "      'AGT',\n",
       "      'AHSG',\n",
       "      'ALB',\n",
       "      'AMBP',\n",
       "      'AOAH',\n",
       "      'APCS',\n",
       "      'APOA1',\n",
       "      'APOA2',\n",
       "      'APOA4',\n",
       "      'APOB',\n",
       "      'APOC1',\n",
       "      'APOC2',\n",
       "      'APOC3',\n",
       "      'APOC4',\n",
       "      'APOD',\n",
       "      'APOE',\n",
       "      'APOF',\n",
       "      'APOH',\n",
       "      'APOL1',\n",
       "      'APOM',\n",
       "      'ATRN',\n",
       "      'AZGP1',\n",
       "      'BCHE',\n",
       "      'BTD',\n",
       "      'C1QA',\n",
       "      'C1QB',\n",
       "      'C1QC',\n",
       "      'C1R',\n",
       "      'C1R;C1RL',\n",
       "      'C1RL',\n",
       "      'C2',\n",
       "      'C3',\n",
       "      'C4A',\n",
       "      'C4A;C4B',\n",
       "      'C4B',\n",
       "      'C4BPA',\n",
       "      'C4BPB',\n",
       "      'C5',\n",
       "      'C6',\n",
       "      'C7',\n",
       "      'C8A',\n",
       "      'C8B',\n",
       "      'C8G',\n",
       "      'C9',\n",
       "      'CD14',\n",
       "      'CD5L',\n",
       "      'CFB',\n",
       "      'CFH',\n",
       "      'CFH;CFHR1',\n",
       "      'CFH;CFHR1;CFHR2',\n",
       "      'CFHR1',\n",
       "      'CFHR1;CFHR2',\n",
       "      'CFHR1;CFHR5',\n",
       "      'CFHR2',\n",
       "      'CFI',\n",
       "      'CFP',\n",
       "      'CLEC3B',\n",
       "      'CLU',\n",
       "      'CNDP1',\n",
       "      'COL26A1',\n",
       "      'CP',\n",
       "      'CPB2',\n",
       "      'CPN1',\n",
       "      'CPN2',\n",
       "      'ECM1',\n",
       "      'EFEMP1',\n",
       "      'ENTPD5',\n",
       "      'F10',\n",
       "      'F11',\n",
       "      'F11;KLKB1',\n",
       "      'F12',\n",
       "      'F13A1',\n",
       "      'F13B',\n",
       "      'F2',\n",
       "      'F5',\n",
       "      'F9',\n",
       "      'FBLN1',\n",
       "      'FCN3',\n",
       "      'FETUB',\n",
       "      'FGA',\n",
       "      'FGB',\n",
       "      'FGG',\n",
       "      'FN1',\n",
       "      'GC',\n",
       "      'GPLD1',\n",
       "      'GPX3',\n",
       "      'GSN',\n",
       "      'HABP2',\n",
       "      'HGFAC',\n",
       "      'HP',\n",
       "      'HP;HPR',\n",
       "      'HPR',\n",
       "      'HPX',\n",
       "      'HRG',\n",
       "      'HTRA4',\n",
       "      'IGFALS',\n",
       "      'IGFBP3',\n",
       "      'IGHA1',\n",
       "      'IGHA1;IGHA2',\n",
       "      'IGHA2',\n",
       "      'IGHD',\n",
       "      'IGHG1',\n",
       "      'IGHG1;IGHG2;IGHG3',\n",
       "      'IGHG1;IGHG2;IGHG3;IGHG4',\n",
       "      'IGHG1;IGHG3',\n",
       "      'IGHG1;IGHG3;IGHG4',\n",
       "      'IGHG1;IGHG4',\n",
       "      'IGHG2',\n",
       "      'IGHG2;IGHG3',\n",
       "      'IGHG2;IGHG3;IGHG4',\n",
       "      'IGHG2;IGHG4',\n",
       "      'IGHG3',\n",
       "      'IGHG4',\n",
       "      'IGHM',\n",
       "      'IGHV1-3;IGHV1-46;IGHV1-69;IGHV1-69D;IGHV1-8',\n",
       "      'IGHV1-45',\n",
       "      'IGHV1-46',\n",
       "      'IGHV1-69',\n",
       "      'IGHV1-69-2;IGHV5-10-1;IGHV5-51',\n",
       "      'IGHV2-26',\n",
       "      'IGHV2-5',\n",
       "      'IGHV2-70D',\n",
       "      'IGHV3-11;IGHV3-13;IGHV3-20;IGHV3-43;IGHV3-7',\n",
       "      'IGHV3-11;IGHV3-33;IGHV3-53;IGHV3-66;IGHV3-7;IGHV3-74',\n",
       "      'IGHV3-13',\n",
       "      'IGHV3-13;IGHV3-23;IGHV3-35;IGHV3-64D;IGHV3-7;IGHV3-72;IGHV3-74',\n",
       "      'IGHV3-15',\n",
       "      'IGHV3-15;IGHV3-72;IGHV3-73',\n",
       "      'IGHV3-23',\n",
       "      'IGHV3-23;IGHV3-33;IGHV3-53;IGHV3-66;IGHV3-74',\n",
       "      'IGHV3-33',\n",
       "      'IGHV3-35',\n",
       "      'IGHV3-43',\n",
       "      'IGHV3-43D;IGHV3-9',\n",
       "      'IGHV3-49',\n",
       "      'IGHV3-49;IGHV3-73',\n",
       "      'IGHV3-53',\n",
       "      'IGHV3-64',\n",
       "      'IGHV3-66',\n",
       "      'IGHV3-7',\n",
       "      'IGHV3-72',\n",
       "      'IGHV3-73',\n",
       "      'IGHV3-74',\n",
       "      'IGHV4-28;IGHV4-38-2;IGHV4-4',\n",
       "      'IGHV4-30-2;IGHV4-34;IGHV4-38-2;IGHV4-4',\n",
       "      'IGHV4-34;IGHV4-38-2',\n",
       "      'IGHV4-4',\n",
       "      'IGHV5-10-1',\n",
       "      'IGHV5-10-1;IGHV5-51',\n",
       "      'IGHV5-51',\n",
       "      'IGHV6-1',\n",
       "      'IGKC',\n",
       "      'IGKJ1',\n",
       "      'IGKV1-12;IGKV1D-12',\n",
       "      'IGKV1-13;IGKV1D-13',\n",
       "      'IGKV1-16',\n",
       "      'IGKV1-16;IGKV1-17;IGKV1-27;IGKV1-39;IGKV1D-16',\n",
       "      'IGKV1-17',\n",
       "      'IGKV1-17;IGKV1-6',\n",
       "      'IGKV1-27;IGKV1-8;IGKV1-9',\n",
       "      'IGKV1-33;IGKV1D-33',\n",
       "      'IGKV1-37;IGKV1D-37',\n",
       "      'IGKV1-39;IGKV1D-39',\n",
       "      'IGKV1-5',\n",
       "      'IGKV1-6',\n",
       "      'IGKV1-9',\n",
       "      'IGKV2-24;IGKV2-29;IGKV2D-24;IGKV2D-29',\n",
       "      'IGKV2-24;IGKV2D-24',\n",
       "      'IGKV2-28;IGKV2D-28',\n",
       "      'IGKV2-29;IGKV2-30;IGKV2D-29',\n",
       "      'IGKV2-29;IGKV2D-29',\n",
       "      'IGKV2-30',\n",
       "      'IGKV2-40;IGKV2D-40',\n",
       "      'IGKV2D-29',\n",
       "      'IGKV3-11',\n",
       "      'IGKV3-11;IGKV3D-20',\n",
       "      'IGKV3-15;IGKV3D-15',\n",
       "      'IGKV3-20',\n",
       "      'IGKV3-20;IGKV3D-20',\n",
       "      'IGKV3-7;IGKV3D-7',\n",
       "      'IGKV3D-20',\n",
       "      'IGKV3D-7',\n",
       "      'IGKV4-1',\n",
       "      'IGKV6-21',\n",
       "      'IGLC3',\n",
       "      'IGLC7',\n",
       "      'IGLC7;IGLL5',\n",
       "      'IGLL5',\n",
       "      'IGLV1-47',\n",
       "      'IGLV1-51',\n",
       "      'IGLV2-11',\n",
       "      'IGLV2-23',\n",
       "      'IGLV2-8',\n",
       "      'IGLV3-1',\n",
       "      'IGLV3-10;IGLV3-16',\n",
       "      'IGLV3-10;IGLV3-25',\n",
       "      'IGLV3-16;IGLV3-25;IGLV3-27',\n",
       "      'IGLV3-19',\n",
       "      'IGLV3-21',\n",
       "      'IGLV3-21;IGLV3-9',\n",
       "      'IGLV3-25;IGLV3-27',\n",
       "      'IGLV3-27',\n",
       "      'IGLV5-45',\n",
       "      'IGLV6-57',\n",
       "      'IGLV7-43;IGLV7-46',\n",
       "      'IGLV8-61',\n",
       "      'ITIH1',\n",
       "      'ITIH2',\n",
       "      'ITIH3',\n",
       "      'ITIH4',\n",
       "      'JCHAIN',\n",
       "      'KLKB1',\n",
       "      'KNG1',\n",
       "      'LBP',\n",
       "      'LCAT',\n",
       "      'LGALS3BP',\n",
       "      'LRG1',\n",
       "      'LUM',\n",
       "      'MASP1',\n",
       "      'MST1',\n",
       "      'ORM1',\n",
       "      'ORM1;ORM2',\n",
       "      'ORM2',\n",
       "      'PF4',\n",
       "      'PGLYRP2',\n",
       "      'PLG',\n",
       "      'PLTP',\n",
       "      'PON1',\n",
       "      'PPBP',\n",
       "      'PRG4',\n",
       "      'PROC',\n",
       "      'PROS1',\n",
       "      'PRSS3',\n",
       "      'PZP',\n",
       "      'RBP4',\n",
       "      'SAA4',\n",
       "      'SELENOP',\n",
       "      'SERPINA1',\n",
       "      'SERPINA10',\n",
       "      'SERPINA3',\n",
       "      'SERPINA4',\n",
       "      'SERPINA5',\n",
       "      'SERPINA6',\n",
       "      'SERPINA7',\n",
       "      'SERPINC1',\n",
       "      'SERPINF1',\n",
       "      'SERPINF2',\n",
       "      'SERPING1',\n",
       "      'SHBG',\n",
       "      'TF',\n",
       "      'TGFBI',\n",
       "      'THBS1',\n",
       "      'TTR',\n",
       "      'VTN']},\n",
       "    'type': 'array'},\n",
       "   'samples': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type', 'submitter_id', 'assay_description'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Proteomics Assay',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'sample.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'biospecimen',\n",
       "  'description': \"Biospecimen information that links subjects to samples including sample's provider and source.\",\n",
       "  'id': 'sample',\n",
       "  'links': [{'backref': 'samples',\n",
       "    'label': 'taken_from',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'name': 'subjects',\n",
       "    'required': True,\n",
       "    'target_type': 'subject'}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'freeze_thaw_cycles': {'description': 'Number of freeze thaw cycles conducted on the sample',\n",
       "    'type': 'integer'},\n",
       "   'sample_collection_method': {'description': 'How the sample was collected, e.g. blood draw',\n",
       "    'type': 'string'},\n",
       "   'sample_id': {'description': 'A unique sample identifier',\n",
       "    'type': 'string'},\n",
       "   'sample_in_preservation': {'description': 'The preservation method used for the sample',\n",
       "    'enum': ['cryopreserved',\n",
       "     'FFPE',\n",
       "     'fresh',\n",
       "     'OCT',\n",
       "     'snap Frozen',\n",
       "     'frozen',\n",
       "     'unknown',\n",
       "     'not reported',\n",
       "     'not allowed to collect'],\n",
       "    'enumDef': [{'enumeration': 'cryopreserved',\n",
       "      'source': 'BioData Catalyst DD'},\n",
       "     {'enumeration': 'FFPE', 'source': 'BioData Catalyst DD'},\n",
       "     {'enumeration': 'fresh', 'source': 'BioData Catalyst DD'},\n",
       "     {'enumeration': 'OCT', 'source': 'BioData Catalyst DD'},\n",
       "     {'enumeration': 'snap Frozen', 'source': 'BioData Catalyst DD'},\n",
       "     {'enumeration': 'frozen', 'source': 'BioData Catalyst DD'},\n",
       "     {'enumeration': 'unknown', 'source': 'BioData Catalyst DD'},\n",
       "     {'enumeration': 'not reported', 'source': 'BioData Catalyst DD'},\n",
       "     {'enumeration': 'not allowed to collect',\n",
       "      'source': 'BioData Catalyst DD'}]},\n",
       "   'sample_in_storage': {'description': 'Whether there is a sample in storage',\n",
       "    'enum': ['yes', 'no', 'unknown']},\n",
       "   'sample_provider': {'description': 'The name of collaborating institute that provided the sample',\n",
       "    'enum': ['Baker', 'USYD', 'UMELB', 'UQ']},\n",
       "   'sample_source': {'description': 'Uberon identifier, anatomical location as described by the Uber Anatomy Ontology (UBERON). (CMG, CCDG)',\n",
       "    'pattern': '^UBERON:[0-9]{7}$',\n",
       "    'type': 'string'},\n",
       "   'sample_storage_method': {'description': 'How the sample was/is stored',\n",
       "    'enum': ['not stored',\n",
       "     'ambient temperature',\n",
       "     'cut slide',\n",
       "     'fresh',\n",
       "     'frozen, -70C freezer',\n",
       "     'frozen, -150C freezer',\n",
       "     'frozen, liquid nitrogen',\n",
       "     'frozen, vapor phase',\n",
       "     'paraffin block',\n",
       "     'RNAlater, frozen',\n",
       "     'TRIzol, frozen'],\n",
       "    'enumDef': [{'enumeration': 'ambient temperature',\n",
       "      'source': 'https://schema.humancellatlas.org/module/biomaterial/6.1.1/preservation_storage',\n",
       "      'version': '6.1.1'},\n",
       "     {'enumeration': 'cut slide',\n",
       "      'source': 'https://schema.humancellatlas.org/module/biomaterial/6.1.1/preservation_storage',\n",
       "      'version': '6.1.1'},\n",
       "     {'enumeration': 'fresh',\n",
       "      'source': 'https://schema.humancellatlas.org/module/biomaterial/6.1.1/preservation_storage',\n",
       "      'version': '6.1.1'},\n",
       "     {'enumeration': 'frozen, -70C freezer',\n",
       "      'source': 'https://schema.humancellatlas.org/module/biomaterial/6.1.1/preservation_storage',\n",
       "      'version': '6.1.1'},\n",
       "     {'enumeration': 'frozen, -150C freezer',\n",
       "      'source': 'https://schema.humancellatlas.org/module/biomaterial/6.1.1/preservation_storage',\n",
       "      'version': '6.1.1'},\n",
       "     {'enumeration': 'frozen, liquid nitrogen',\n",
       "      'source': 'https://schema.humancellatlas.org/module/biomaterial/6.1.1/preservation_storage',\n",
       "      'version': '6.1.1'},\n",
       "     {'enumeration': 'frozen, vapor phase',\n",
       "      'source': 'https://schema.humancellatlas.org/module/biomaterial/6.1.1/preservation_storage',\n",
       "      'version': '6.1.1'},\n",
       "     {'enumeration': 'paraffin block',\n",
       "      'source': 'https://schema.humancellatlas.org/module/biomaterial/6.1.1/preservation_storage',\n",
       "      'version': '6.1.1'},\n",
       "     {'enumeration': 'RNAlater, frozen',\n",
       "      'source': 'https://schema.humancellatlas.org/module/biomaterial/6.1.1/preservation_storage',\n",
       "      'version': '6.1.1'},\n",
       "     {'enumeration': 'TRIzol, frozen',\n",
       "      'source': 'https://schema.humancellatlas.org/module/biomaterial/6.1.1/preservation_storage',\n",
       "      'version': '6.1.1'}]},\n",
       "   'sample_type': {'description': 'The type of sample, e.g. Whole Blood, Blood Cells, Frozen Blood',\n",
       "    'type': 'string'},\n",
       "   'storage_location': {'description': 'Where the sample is stored',\n",
       "    'enum': ['Baker', 'USYD', 'UMELB', 'UQ']},\n",
       "   'subjects': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Sample',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'unaligned_reads_file.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'data_file',\n",
       "  'description': 'Data file containing raw reads from a sequencing experiment.',\n",
       "  'id': 'unaligned_reads_file',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'unaligned_reads_files',\n",
       "      'label': 'generated_from',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'genomics_assay',\n",
       "      'required': False,\n",
       "      'target_type': 'genomics_assay'},\n",
       "     {'backref': 'aligned_reads_filess',\n",
       "      'label': 'data_from',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'core_metadata_collections',\n",
       "      'required': False,\n",
       "      'target_type': 'core_metadata_collection'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/data_file_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'core_metadata_collections': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'data_category': {'description': 'Broad categorization of the contents of the data file.',\n",
       "    'enum': ['analysis',\n",
       "     'sequencing reads',\n",
       "     'single nucleotide variation',\n",
       "     'transcriptome profiling',\n",
       "     'supplemental']},\n",
       "   'data_format': {'description': 'Format of the data files.',\n",
       "    'enum': ['fastq.gz']},\n",
       "   'data_type': {'description': 'Specific content type of the data file.',\n",
       "    'enum': ['unaligned reads',\n",
       "     'aligned reads',\n",
       "     'variants annotation',\n",
       "     'clinical supplement']},\n",
       "   'genomics_assay': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'run_id': {'description': 'Sequencing run ID associated with file',\n",
       "    'type': 'string'}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'data_category',\n",
       "   'data_format',\n",
       "   'data_type'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Unaligned Reads File',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " '_terms.yaml': {'28s_16s_ribosomal_rna_ratio': {'description': 'The 28S/18S ribosomal RNA band ratio used to assess the quality of total RNA.\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': '28s/18s Ribosomal RNA Ratio',\n",
       "    'term_url': None}},\n",
       "  'RIN': {'description': 'A numerical assessment of the integrity of RNA based on the entire electrophoretic trace of the RNA sample including the presence or absence of degradation products.\\n',\n",
       "   'termDef': {'cde_id': 5278775,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Biospecimen RNA Integrity Number Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5278775&version=1.0'}},\n",
       "  'UUID': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "   'termDef': {'cde_id': 'C54100',\n",
       "    'cde_version': None,\n",
       "    'source': 'NCIt',\n",
       "    'term': 'Universally Unique Identifier',\n",
       "    'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "  'a260_a280_ratio': {'description': 'Numeric value that represents the sample ratio of nucleic acid absorbance at 260 nm and 280 nm, used to determine a measure of DNA purity.\\n',\n",
       "   'termDef': {'cde_id': 5432595,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Nucleic Acid Absorbance at 260 And Absorbance at 280 DNA Purity Ratio Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432595&version=1.0'}},\n",
       "  'adapter_name': {'description': 'Name of the sequencing adapter.\\n'},\n",
       "  'adapter_sequence': {'description': 'Base sequence of the sequencing adapter.\\n'},\n",
       "  'age': {'description': 'Age at last visit. Age is encoded as ISO8601 duration. (CMG)\\n',\n",
       "   'termDef': {'cde_id': 'ISO8601',\n",
       "    'cde_version': 1,\n",
       "    'source': 'Phenopackets',\n",
       "    'term': 'Age at last visit. Age is encoded as ISO8601 duration.',\n",
       "    'term_url': 'https://phenopackets-schema.readthedocs.io/en/latest/age.html#rstage'}},\n",
       "  'age_at_diagnosis': {'description': \"Age at the time of diagnosis expressed in number of days since birth. If the age at diagnosis is greater than 32872 days (89 years), then see 'age_at_diagnosis_gt89'.\\n\",\n",
       "   'termDef': {'cde_id': 3225640,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Patient Diagnosis Age Day Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3225640&version=2.0'}},\n",
       "  'age_at_diagnosis_gt89': {'description': 'Indicate whether the age at the time of diagnosis expressed in number of days since birth is greater than 32872 days (89 years).\\n',\n",
       "   'termDef': {'cde_id': 3225640,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Patient Diagnosis Age Day Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3225640&version=2.0'}},\n",
       "  'age_at_event_days_gt89': {'description': \"Indicate whether the 'Age at the time an event occurs expressed in the number of days since birth' is greater than 32872 days (89 years).\\n\"},\n",
       "  'age_of_onset': {'description': 'Age of onset (if affected). Otherwise age of last exam or phenotype transfer. (CMG)\\n',\n",
       "   'termDef': {'cde_id': 'ISO8601',\n",
       "    'cde_version': 1,\n",
       "    'source': 'Phenopackets',\n",
       "    'term': 'Age of onset (if affected). Otherwise age of last exam or phenotype transfer. (CMG)',\n",
       "    'term_url': 'https://phenopackets-schema.readthedocs.io/en/latest/age.html#rstage'}},\n",
       "  'ajcc_clinical_m': {'description': 'Extent of the distant metastasis for the cancer based on evidence obtained from clinical assessment parameters determined prior to treatment.\\n',\n",
       "   'termDef': {'cde_id': 3440331,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Neoplasm American Joint Committee on Cancer Clinical Distant Metastasis M Stage',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3440331&version=1.0'}},\n",
       "  'ajcc_clinical_n': {'description': 'Extent of the regional lymph node involvement for the cancer based on evidence obtained from clinical assessment parameters determined prior to treatment.\\n',\n",
       "   'termDef': {'cde_id': 3440330,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Neoplasm American Joint Committee on Cancer Clinical Regional Lymph Node N Stage',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3440330&version=1.0'}},\n",
       "  'ajcc_clinical_stage': {'description': 'Stage group determined from clinical information on the tumor (T), regional node (N) and metastases (M) and by grouping cases with similar prognosis for cancer.\\n',\n",
       "   'termDef': {'cde_id': 3440332,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Neoplasm American Joint Committee on Cancer Clinical Group Stage',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3440332&version=1.0'}},\n",
       "  'ajcc_clinical_t': {'description': 'Extent of the primary cancer based on evidence obtained from clinical assessment parameters determined prior to treatment.\\n',\n",
       "   'termDef': {'cde_id': 3440328,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Neoplasm American Joint Committee on Cancer Clinical Primary Tumor T Stage',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3440328&version=1.0'}},\n",
       "  'ajcc_pathologic_m': {'description': 'Code to represent the defined absence or presence of distant spread or metastases (M) to locations via vascular channels or lymphatics beyond the regional lymph nodes, using criteria established by the American Joint Committee on Cancer (AJCC).\\n',\n",
       "   'termDef': {'cde_id': 3045439,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'American Joint Committee on Cancer Metastasis Stage Code',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3045439&version=1.0'}},\n",
       "  'ajcc_pathologic_n': {'description': \"The codes that represent the stage of cancer based on the nodes present (N stage) according to criteria based on multiple editions of the AJCC's Cancer Staging Manual.\\n\",\n",
       "   'termDef': {'cde_id': 3203106,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Neoplasm Disease Lymph Node Stage American Joint Committee on Cancer Code',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3203106&version=1.0'}},\n",
       "  'ajcc_pathologic_stage': {'description': 'The extent of a cancer, especially whether the disease has spread from the original site to other parts of the body based on AJCC staging criteria.\\n',\n",
       "   'termDef': {'cde_id': 3203222,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Neoplasm Disease Stage American Joint Committee on Cancer Code',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3203222&version=1.0'}},\n",
       "  'ajcc_pathologic_t': {'description': 'Code of pathological T (primary tumor) to define the size or contiguous extension of the primary tumor (T), using staging criteria from the American Joint Committee on Cancer (AJCC).\\n',\n",
       "   'termDef': {'cde_id': 3045435,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'American Joint Committee on Cancer Tumor Stage Code',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3045435&version=1.0'}},\n",
       "  'alcohol_history': {'description': 'A response to a question that asks whether the participant has consumed at least 12 drinks of any kind of alcoholic beverage in their lifetime.\\n',\n",
       "   'termDef': {'cde_id': 2201918,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Alcohol Lifetime History Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2201918&version=1.0'}},\n",
       "  'aliquot_quantity': {'description': 'The quantity in micrograms (ug) of the aliquot(s) derived from the analyte(s) shipped for sequencing and characterization.\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': 'Biospecimen Aliquot Quantity',\n",
       "    'term_url': None}},\n",
       "  'aliquot_volume': {'description': 'The volume in microliters (ml) of the aliquot(s) derived from the analyte(s) shipped for sequencing and characterization.\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': 'Biospecimen Aliquot Volume',\n",
       "    'term_url': None}},\n",
       "  'amount': {'description': 'Weight in grams or volume in mL.\\n'},\n",
       "  'analyte_quantity': {'description': 'The quantity in micrograms (ug) of the analyte(s) derived from the analyte(s) shipped for sequencing and characterization.\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': 'Biospecimen Analyte Quantity',\n",
       "    'term_url': None}},\n",
       "  'analyte_type': {'description': 'Text term that represents the kind of molecular specimen analyte.\\n',\n",
       "   'termDef': {'cde_id': 2513915,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Molecular Specimen Type Text Name',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2513915&version=2.0'}},\n",
       "  'analyte_type_id': {'description': 'A single letter code used to identify a type of molecular analyte.\\n',\n",
       "   'termDef': {'cde_id': 5432508,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Molecular Analyte Identification Code',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432508&version=1.0'}},\n",
       "  'analyte_volume': {'description': 'The volume in microliters (ml) of the analyte(s) derived from the analyte(s) shipped for sequencing and characterization.\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': 'Biospecimen Analyte Volume',\n",
       "    'term_url': None}},\n",
       "  'ann_arbor_b_symptoms': {'description': \"Text term to signify whether lymphoma B-symptoms are present as noted in the patient's medical record.\\n\",\n",
       "   'termDef': {'cde_id': 2902402,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Lymphoma B-Symptoms Medical Record Documented Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2902402&version=1.0'}},\n",
       "  'ann_arbor_clinical_stage': {'description': \"The classification of the clinically confirmed anatomic disease extent of lymphoma (Hodgkin's and Non-Hodgkins) based on the Ann Arbor Staging System.\\n\",\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': 'Ann Arbor Clinical Stage',\n",
       "    'term_url': None}},\n",
       "  'ann_arbor_extranodal_involvement': {'description': 'Indicator that identifies whether a patient with malignant lymphoma has lymphomatous involvement of an extranodal site.\\n',\n",
       "   'termDef': {'cde_id': 3364582,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Lymphomatous Extranodal Site Involvement Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3364582&version=1.0'}},\n",
       "  'ann_arbor_pathologic_stage': {'description': \"The classification of the pathologically confirmed anatomic disease extent of lymphoma (Hodgkin's and Non-Hodgkins) based on the Ann Arbor Staging System.\\n\",\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': 'Ann Arbor Pathologic Stage',\n",
       "    'term_url': None}},\n",
       "  'ann_arbor_tumor_stage': {'description': \"The classification of the anatomic disease extent of lymphoma (Hodgkin's and Non-Hodgkins) based on the Ann Arbor Staging System.\\n\",\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': 'Ann Arbor Tumor Stage',\n",
       "    'term_url': None}},\n",
       "  'base_caller_name': {'description': 'Name of the base caller.\\n'},\n",
       "  'base_caller_version': {'description': 'Version of the base caller.\\n'},\n",
       "  'biomarker_name': {'description': 'The name of the biomarker being tested for this specimen and set of test results.\\n',\n",
       "   'termDef': {'cde_id': 5473,\n",
       "    'cde_version': 11,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Biomarker Name',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5473&version=2.31'}},\n",
       "  'biomarker_result': {'description': 'Text term to define the results of genetic testing.\\n',\n",
       "   'termDef': {'cde_id': 3234680,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Laboratory Procedure Genetic Abnormality Test Result Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3234680&version=1.0'}},\n",
       "  'biomarker_test_method': {'description': 'Text descriptor of a molecular analysis method used for an individual.\\n',\n",
       "   'termDef': {'cde_id': 3121575,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Disease Detection Molecular Analysis Method Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3121575&version=1.0'}},\n",
       "  'biospecimen_anatomic_site_id': {'description': 'Uberon ID, anatomical location as described by the Uber Anatomy Ontology (UBERON). (CMG)\\n',\n",
       "   'termDef': {'cde_id': 'Uberon',\n",
       "    'cde_version': 1,\n",
       "    'source': 'Uberon',\n",
       "    'term': 'Uberon ID, anatomical location as described by the Uber Anatomy Ontology (UBERON). (CMG)',\n",
       "    'term_url': 'https://www.ebi.ac.uk/ols/ontologies/uberon'}},\n",
       "  'biospecimen_anatomic_site_label': {'description': 'Uberon Label, anatomical location as described by the Uber Anatomy Ontology (UBERON). (CMG)\\n',\n",
       "   'termDef': {'cde_id': 'Uberon',\n",
       "    'cde_version': 1,\n",
       "    'source': 'Uberon',\n",
       "    'term': 'Uberon Label, anatomical location as described by the Uber Anatomy Ontology (UBERON). (CMG)',\n",
       "    'term_url': 'https://www.ebi.ac.uk/ols/ontologies/uberon'}},\n",
       "  'bmi': {'description': 'Autocalculated field of BMI: general indicator of the body fat an individual is carrying based upon the ratio of weight to height. BMI = 703 * weight in pounds/height in inches squared. (GTEx)\\n',\n",
       "   'termDef': {'cde_id': 4973892,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Body Mass Index (BMI)',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4973892&version=1.0'}},\n",
       "  'burkitt_lymphoma_clinical_variant': {'description': \"Burkitt's lymphoma categorization based on clinical features that differ from other forms of the same disease.\\n\",\n",
       "   'termDef': {'cde_id': 3770421,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Burkitt Lymphoma Clinical Variant Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3770421&version=1.0'}},\n",
       "  'cause_of_death': {'description': 'Text term to identify the cause of death for a patient.\\n',\n",
       "   'termDef': {'cde_id': 2554674,\n",
       "    'cde_version': 3,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Patient Death Reason',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2554674&version=3.0'}},\n",
       "  'cea_level_preoperative': {'description': 'Numeric value of the Carcinoembryonic antigen or CEA at the time before surgery. [Manually- curated]\\n',\n",
       "   'termDef': {'cde_id': 2716510,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Preoperative Carcinoembryonic Antigen Result Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2716510&version=1.0'}},\n",
       "  'cigarettes_per_day': {'description': 'The average number of cigarettes smoked per day.\\n',\n",
       "   'termDef': {'cde_id': 2001716,\n",
       "    'cde_version': 4,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Smoking Use Average Number',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2001716&version=4.0'}},\n",
       "  'circumferential_resection_margin': {'description': 'A value in millimeters indicating the measured length between a malignant lesion of the colon or rectum and the nearest radial (or circumferential) border of tissue removed during cancer surgery.\\n',\n",
       "   'termDef': {'cde_id': 64202,\n",
       "    'cde_version': 3,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Colorectal Surgical Margin Circumferential Distance Measurement',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=64202&version=3.0'}},\n",
       "  'classification_of_tumor': {'description': 'Text that describes the kind of disease present in the tumor specimen as related to a specific timepoint.\\n',\n",
       "   'termDef': {'cde_id': 3288124,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tumor Tissue Disease Description Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3288124&version=1.0'}},\n",
       "  'colon_polyps_history': {'description': 'Yes/No indicator to describe if the subject had a previous history of colon polyps as noted in the history/physical or previous endoscopic report (s).\\n',\n",
       "   'termDef': {'cde_id': 3107197,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Colon Carcinoma Polyp Occurrence Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3107197&version=1.0'}},\n",
       "  'composition': {'description': 'Text term that represents the cellular composition of the sample.\\n',\n",
       "   'termDef': {'cde_id': 5432591,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Biospecimen Cellular Composition Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432591&version=1.0'}},\n",
       "  'concentration': {'description': 'Numeric value that represents the concentration of an analyte or aliquot extracted from the sample or sample portion, measured in milligrams per milliliter.\\n',\n",
       "   'termDef': {'cde_id': 5432594,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Biospecimen Analyte or Aliquot Extracted Concentration Milligram per Milliliter Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432594&version=1.0'}},\n",
       "  'consent_codes': {'description': 'Data Use Restrictions that are used to indicate  permissions/restrictions for datasets and/or materials, and relates to the purposes for which datasets and/or material might be removed, stored or used. Based on the Data Use Ontology : see http://www.obofoundry.org/ontology/duo.html'},\n",
       "  'creation_datetime': {'description': 'The datetime of portion creation encoded as seconds from epoch.\\n',\n",
       "   'termDef': {'cde_id': 5432592,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Biospecimen Portion Creation Seconds Date/Time',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432592&version=1.0'}},\n",
       "  'current_weight': {'description': 'Numeric value that represents the current weight of the sample, measured  in milligrams.\\n',\n",
       "   'termDef': {'cde_id': 5432606,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tissue Sample Current Weight Milligram Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432606&version=1.0'}},\n",
       "  'data_category': {'description': 'Broad categorization of the contents of the data file.\\n'},\n",
       "  'data_file_error_type': {'description': 'Type of error for the data file object.\\n'},\n",
       "  'data_format': {'description': 'Format of the data files.\\n'},\n",
       "  'data_type': {'description': 'Specific content type of the data file. (CMG)\\n'},\n",
       "  'datetime': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'},\n",
       "  'days_to_birth': {'description': \"Time interval from a person's date of birth to the date of initial pathologic diagnosis, represented as a calculated negative number of days.\\n\",\n",
       "   'termDef': {'cde_id': 3008233,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Person Birth Date Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3008233&version=1.0'}},\n",
       "  'days_to_collection': {'description': 'Time interval from the date of biospecimen collection to the date of initial pathologic diagnosis, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'cde_id': 3008340,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Biospecimen Collection Date Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3008340&version=1.0'}},\n",
       "  'days_to_death': {'description': \"Time interval from a person's date of death to the date of initial pathologic diagnosis, represented as a calculated number of days.\\n\",\n",
       "   'termDef': {'cde_id': 3165475,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Death Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3165475&version=1.0'}},\n",
       "  'days_to_hiv_diagnosis': {'description': 'Time interval from the date of the initial pathologic diagnosis to the date of human immunodeficiency diagnosis, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'cde_id': 4618491,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Human Immunodeficiency Virus Diagnosis Subtract Initial Pathologic Diagnosis Time Duration Day Calculation Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4618491&version=1.0'}},\n",
       "  'days_to_last_follow_up': {'description': 'Time interval from the date of last follow up to the date of initial pathologic diagnosis, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'cde_id': 3008273,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Last Communication Contact Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3008273&version=1.0'}},\n",
       "  'days_to_last_known_disease_status': {'description': 'Time interval from the date of last follow up to the date of initial pathologic diagnosis, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'cde_id': 3008273,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Last Communication Contact Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3008273&version=1.0'}},\n",
       "  'days_to_new_event': {'description': 'Time interval from the date of new tumor event including progression, recurrence and new primary malignacies to the date of initial pathologic diagnosis, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'cde_id': 3392464,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'New Tumor Event Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3392464&version=1.0'}},\n",
       "  'days_to_recurrence': {'description': 'Time interval from the date of new tumor event including progression, recurrence and new primary malignancies to the date of initial pathologic diagnosis, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'cde_id': 3392464,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'New Tumor Event Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3392464&version=1.0'}},\n",
       "  'days_to_sample_procurement': {'description': 'The number of days from the date the patient was diagnosed to the date of the procedure that produced the sample.\\n'},\n",
       "  'days_to_treatment': {'description': 'Number of days from date of initial pathologic diagnosis that treatment began.\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': 'Days to Treatment Start',\n",
       "    'term_url': None}},\n",
       "  'days_to_treatment_end': {'description': 'Time interval from the date of the initial pathologic diagnosis to the date of treatment end, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'cde_id': 5102431,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Treatment End Subtract First Pathologic Diagnosis Day Calculation Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5102431&version=1.0'}},\n",
       "  'days_to_treatment_start': {'description': 'Time interval from the date of the initial pathologic diagnosis to the start of treatment, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'cde_id': 5102411,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Treatment Start Subtract First Pathologic Diagnosis Time Day Calculation Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5102411&version=1.0'}},\n",
       "  'diagnosis_pathologically_confirmed': {'description': 'The histologic description of tissue or cells confirmed by a pathology review of frozen or formalin fixed slide(s) completed after the diagnostic pathology review of the tumor sample used to extract analyte(s).\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': 'Post-Diagnostic Pathology Review Confirmation',\n",
       "    'term_url': None}},\n",
       "  'disease_omim_id': {'description': 'The ID of the term from OMIM which represents a disease. (CMG)\\n',\n",
       "   'termDef': {'cde_id': 'OMIM',\n",
       "    'cde_version': 1,\n",
       "    'source': 'OMIM',\n",
       "    'term': 'The ID of the term from OMIM which represents a disease. (CMG)',\n",
       "    'term_url': 'https://www.ncbi.nlm.nih.gov/omim'}},\n",
       "  'dlco_ref_predictive_percent': {'description': \"The value, as a percentage of predicted lung volume, measuring the amount of carbon monoxide detected in a patient's lungs.\\n\",\n",
       "   'termDef': {'cde_id': 2180255,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Lung Carbon Monoxide Diffusing Capability Test Assessment Predictive Value Percentage Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2180255&version=1.0'}},\n",
       "  'drink_number': {'description': 'Units of drink consumed during Drink Period (DRNKPRD). (GTEx)\\n'},\n",
       "  'encoding': {'description': 'Version of ASCII encoding of quality values found in the file.\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': 'FastQC',\n",
       "    'term': 'Encoding',\n",
       "    'term_url': 'http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/1%20Basic%20Statistics.html'}},\n",
       "  'estrogen_receptor_percent_positive_ihc': {'description': 'Classification to represent ER Positive results expressed as a percentage value.\\n',\n",
       "   'termDef': {'cde_id': 3128341,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'ER Level Cell Percentage Category',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3128341&version=1.0'}},\n",
       "  'estrogen_receptor_result_ihc': {'description': 'Text term to represent the overall result of Estrogen Receptor (ER) testing.\\n',\n",
       "   'termDef': {'cde_id': 2957359,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Breast Carcinoma Estrogen Receptor Status',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2957359&version=2.0'}},\n",
       "  'ethnicity': {'description': \"An individual's self-described social and cultural grouping, specifically whether an individual describes themselves as Hispanic or Latino (HARMONIZED). The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau. (CMG, GTEx)\\n\",\n",
       "   'termDef': {'cde_id': 2192217,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Ethnic Group Category Text',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2192217&version=2.0'}},\n",
       "  'experiment_name': {'description': 'Submitter-defined name for the experiment.\\n'},\n",
       "  'experimental_strategy': {'description': 'The sequencing strategy used to generate the data file referred to by the ga4gh_drs_uri.\\n'},\n",
       "  'family_onset': {'description': 'Typical age of onset for the family. Age is encoded as ISO8601 duration. (CMG)\\n',\n",
       "   'termDef': {'cde_id': 'ISO8601',\n",
       "    'cde_version': 1,\n",
       "    'source': 'Phenopackets',\n",
       "    'term': 'Typical age of onset for the family. Age is encoded as ISO8601 duration.',\n",
       "    'term_url': 'https://phenopackets-schema.readthedocs.io/en/latest/age.html#rstage'}},\n",
       "  'fastq_name': {'description': 'Names of FASTQs.\\n'},\n",
       "  'fev1_fvc_post_bronch_percent': {'description': 'Percentage value to represent result of Forced Expiratory Volume in 1 second (FEV1) divided by the Forced Vital Capacity (FVC) post-bronchodilator.\\n',\n",
       "   'termDef': {'cde_id': 3302956,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Post Bronchodilator FEV1/FVC Percent Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3302956&version=1.0'}},\n",
       "  'fev1_fvc_pre_bronch_percent': {'description': 'Percentage value to represent result of Forced Expiratory Volume in 1 second (FEV1) divided by the Forced Vital Capacity (FVC) pre-bronchodilator.\\n',\n",
       "   'termDef': {'cde_id': 3302955,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Pre Bronchodilator FEV1/FVC Percent Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3302955&version=1.0'}},\n",
       "  'fev1_ref_post_bronch_percent': {'description': 'The percentage comparison to a normal value reference range of the volume of air that a patient can forcibly exhale from the lungs in one second post-bronchodilator.\\n',\n",
       "   'termDef': {'cde_id': 3302948,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Post Bronchodilator Lung Forced Expiratory Volume 1 Test Lab Percentage Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3302948&version=1.0'}},\n",
       "  'fev1_ref_pre_bronch_percent': {'description': 'The percentage comparison to a normal value reference range of the volume of air that a patient can forcibly exhale from the lungs in one second pre-bronchodilator.\\n',\n",
       "   'termDef': {'cde_id': 3302947,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Pre Bronchodilator Lung Forced Expiratory Volume 1 Test Lab Percentage Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3302947&version=1.0'}},\n",
       "  'figo_stage': {'description': 'The extent of a cervical or endometrial cancer within the body, especially whether the disease has spread from the original site to other parts of the body, as described by the International Federation of Gynecology and Obstetrics (FIGO) stages.\\n',\n",
       "   'termDef': {'cde_id': 3225684,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Gynecologic Tumor Grouping Cervical Endometrial FIGO 2009 Stage',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3225684&version=1.0'}},\n",
       "  'file_format': {'description': 'The format of the data file object.\\n'},\n",
       "  'file_name': {'description': 'The name (or part of a name) of a file (of any type).\\n'},\n",
       "  'file_size': {'description': 'The size of the data file (object) in bytes.\\n'},\n",
       "  'file_state': {'description': 'The current state of the data file object.\\n'},\n",
       "  'flow_cell_barcode': {'description': 'Flow Cell Barcode.\\n'},\n",
       "  'freezing_method': {'description': 'Text term that represents the method used for freezing the sample.\\n',\n",
       "   'termDef': {'cde_id': 5432607,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tissue Sample Freezing Method Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432607&version=1.0'}},\n",
       "  'ga4gh_drs_uri': {'description': 'DRS URI as defined by GA4GH DRS spec for pointers to file objects.\\n'},\n",
       "  'height': {'description': 'The height of Donor in inches as reported by Donor, family/next of kin, or medical record abstraction. (GTEx)\\n',\n",
       "   'termDef': {'cde_id': 649,\n",
       "    'cde_version': 4.1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Patient Height Measurement',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=649&version=4.1'}},\n",
       "  'her2_erbb2_percent_positive_ihc': {'description': 'Classification to represent the number of positive HER2/ERBB2 cells in a specimen or sample.\\n',\n",
       "   'termDef': {'cde_id': 3086980,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'HER2 ERBB Positive Finding Cell Percentage Category',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3086980&version=1.0'}},\n",
       "  'her2_erbb2_result_fish': {'description': 'the type of outcome for HER2 as determined by an in situ hybridization (ISH) assay.\\n',\n",
       "   'termDef': {'cde_id': 2854089,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Laboratory Procedure HER2/neu in situ Hybridization Outcome Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2854089&version=1.0'}},\n",
       "  'her2_erbb2_result_ihc': {'description': 'Text term to signify the result of the medical procedure that involves testing a sample of blood or tissue for HER2 by histochemical localization of immunoreactive substances using labeled antibodies as reagents.\\n',\n",
       "   'termDef': {'cde_id': 2957563,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Laboratory Procedure HER2/neu Immunohistochemistry Receptor Status',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2957563&version=2.0'}},\n",
       "  'hiv_positive': {'description': 'Text term to signify whether a physician has diagnosed HIV infection in a patient.\\n',\n",
       "   'termDef': {'cde_id': 4030799,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Physician Diagnosed HIV Infection Personal Medical History Yes No Not Applicable Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4030799&version=1.0'}},\n",
       "  'hpv_positive_type': {'description': 'Text classification to represent the strain or type of human papillomavirus identified in an individual.\\n',\n",
       "   'termDef': {'cde_id': 2922649,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Human Papillomavirus Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2922649&version=1.0'}},\n",
       "  'hpv_status': {'description': 'The findings of the oncogenic HPV.\\n',\n",
       "   'termDef': {'cde_id': 2230033,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Oncogenic Human Papillomavirus Result Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2230033&version=1.0'}},\n",
       "  'id': '_terms',\n",
       "  'includes_spike_ins': {'description': 'Spike-in included?\\n'},\n",
       "  'initial_weight': {'description': 'Numeric value that represents the initial weight of the sample, measured in milligrams.\\n',\n",
       "   'termDef': {'cde_id': 5432605,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tissue Sample Initial Weight Milligram Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432605&version=1.0'}},\n",
       "  'instrument_model': {'description': 'Numeric value that represents the sample dimension that is greater than the shortest dimension and less than the longest dimension, measured in millimeters.\\n',\n",
       "   'termDef': {'cde_id': 5432604,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tissue Sample Intermediate Dimension Millimeter Measurement',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432604&version=1.0'}},\n",
       "  'intermediate_dimension': {'description': 'Intermediate dimension of the sample, in millimeters.\\n'},\n",
       "  'is_ffpe': {'description': 'Indicator to signify whether or not the tissue sample was fixed in formalin and embedded in paraffin (FFPE).\\n',\n",
       "   'termDef': {'cde_id': 4170557,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Specimen Processing Formalin Fixed Paraffin Embedded Tissue Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4170557&version=1.0'}},\n",
       "  'is_paired_end': {'description': 'Are the reads paired end?\\n'},\n",
       "  'is_proband': {'description': 'Boolean value to determine if the case is the proband.\\n'},\n",
       "  'last_known_disease_status': {'description': \"Text term that describes the last known state or condition of an individual's neoplasm.\\n\",\n",
       "   'termDef': {'cde_id': 5424231,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Person Last Known Neoplasm Status',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2759550&version=1.0'}},\n",
       "  'laterality': {'description': 'For tumors in paired organs, designates the side on which the cancer originates.\\n',\n",
       "   'termDef': {'cde_id': 827,\n",
       "    'cde_version': 3,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Primary Tumor Laterality',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=827&version=3.0'}},\n",
       "  'ldh_level_at_diagnosis': {'description': 'The 2 decimal place numeric laboratory value measured, assigned or computed related to the assessment of lactate dehydrogenase in a specimen.\\n',\n",
       "   'termDef': {'cde_id': 2798766,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Laboratory Procedure Lactate Dehydrogenase Result Integer::2 Decimal Place Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2798766&version=1.0'}},\n",
       "  'ldh_normal_range_upper': {'description': 'The top value of the range of statistical characteristics that are supposed to represent accepted standard, non-pathological pattern for lactate dehydrogenase (units not specified).\\n',\n",
       "   'termDef': {'cde_id': 2597015,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Laboratory Procedure Lactate Dehydrogenase Result Upper Limit of Normal Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2597015&version=1.0'}},\n",
       "  'library_name': {'description': 'Name of the library.\\n'},\n",
       "  'library_preparation_kit_catalog_number': {'description': 'Catalog of Library Preparation Kit\\n'},\n",
       "  'library_preparation_kit_name': {'description': 'Name of Library Preparation Kit\\n'},\n",
       "  'library_preparation_kit_vendor': {'description': 'Vendor of Library Preparation Kit\\n'},\n",
       "  'library_preparation_kit_version': {'description': 'Version of Library Preparation Kit\\n'},\n",
       "  'library_selection': {'description': 'Library Selection Method\\n'},\n",
       "  'library_strand': {'description': 'Library stranded-ness.\\n'},\n",
       "  'library_strategy': {'description': 'Library strategy.\\n'},\n",
       "  'longest_dimension': {'description': 'Numeric value that represents the longest dimension of the sample, measured in millimeters.\\n',\n",
       "   'termDef': {'cde_id': 5432602,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tissue Sample Longest Dimension Millimeter Measurement',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432602&version=1.0'}},\n",
       "  'lymph_nodes_positive': {'description': 'The number of lymph nodes involved with disease as determined by pathologic examination.\\n',\n",
       "   'termDef': {'cde_id': 89,\n",
       "    'cde_version': 3,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Lymph Node(s) Positive Number',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=89&version=3.0'}},\n",
       "  'lymphatic_invasion_present': {'description': 'A yes/no indicator to ask if small or thin-walled vessel invasion is present, indicating lymphatic involvement\\n',\n",
       "   'termDef': {'cde_id': 64171,\n",
       "    'cde_version': 3,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Lymphatic/Small vessel Invasion Ind',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=64171&version=3.0'}},\n",
       "  'md5sum': {'description': \"The 128-bit hash value expressed as a 32 digit hexadecimal number used as a file's digital fingerprint.\\n\"},\n",
       "  'method_of_diagnosis': {'description': \"The method used to initially the patient's diagnosis.\\n\",\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': 'Method of Diagnosis',\n",
       "    'term_url': None}},\n",
       "  'method_of_sample_procurement': {'description': 'The method used to procure the sample used to extract analyte(s).\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': 'Method of Sample Procurement',\n",
       "    'term_url': None}},\n",
       "  'microsatellite_instability_abnormal': {'description': 'The yes/no indicator to signify the status of a tumor for microsatellite instability.\\n',\n",
       "   'termDef': {'cde_id': 3123142,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Microsatellite Instability Occurrence Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3123142&version=1.0'}},\n",
       "  'morphology': {'description': 'The third edition of the International Classification of Diseases for Oncology, published in 2000 used principally in tumor and cancer registries for coding the site (topography) and the histology (morphology) of neoplasms. The study of the structure of the cells and their arrangement to constitute tissues and, finally, the association among these to form organs. In pathology, the microscopic process of identifying normal and abnormal morphologic characteristics in tissues, by employing various cytochemical and immunocytochemical stains. A system of numbered categories for representation of data.\\n',\n",
       "   'termDef': {'cde_id': 3226275,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'International Classification of Diseases for Oncology, Third Edition ICD-O-3 Histology Code',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3226275&version=1.0'}},\n",
       "  'new_event_anatomic_site': {'description': 'Text term to specify the anatomic location of the return of tumor after treatment.\\n',\n",
       "   'termDef': {'cde_id': 3108271,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'New Neoplasm Event Occurrence Anatomic Site',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3108271&version=2.0'}},\n",
       "  'new_event_type': {'description': 'Text term to identify a new tumor event.\\n',\n",
       "   'termDef': {'cde_id': 3119721,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'New Neoplasm Event Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3119721&version=1.0'}},\n",
       "  'normal_tumor_genotype_snp_match': {'description': 'Text term that represents whether or not the genotype of the normal tumor matches or if the data is not available.\\n',\n",
       "   'termDef': {'cde_id': 4588156,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Normal Tumor Genotype Match Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4588156&version=1.0'}},\n",
       "  'number_proliferating_cells': {'description': 'Numeric value that represents the count of proliferating cells determined during pathologic review of the sample slide(s).\\n',\n",
       "   'termDef': {'cde_id': 5432636,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Pathology Review Slide Proliferating Cell Count',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432636&version=1.0'}},\n",
       "  'oct_embedded': {'description': 'Indicator of whether or not the sample was embedded in Optimal Cutting Temperature (OCT) compound.\\n',\n",
       "   'termDef': {'cde_id': 5432538,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tissue Sample Optimal Cutting Temperature Compound Embedding Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432538&version=1.0'}},\n",
       "  'percent_eosinophil_infiltration': {'description': 'Numeric value to represent the percentage of infiltration by eosinophils in a tumor sample or specimen.\\n',\n",
       "   'termDef': {'cde_id': 2897700,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Specimen Eosinophilia Percentage Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2897700&version=2.0'}},\n",
       "  'percent_gc_content': {'description': 'The overall %GC of all bases in all sequences.\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': 'FastQC',\n",
       "    'term': '%GC',\n",
       "    'term_url': 'http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/1%20Basic%20Statistics.html'}},\n",
       "  'percent_granulocyte_infiltration': {'description': 'Numeric value to represent the percentage of infiltration by granulocytes in a tumor sample or specimen.\\n',\n",
       "   'termDef': {'cde_id': 2897705,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Specimen Granulocyte Infiltration Percentage Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2897705&version=2.0'}},\n",
       "  'percent_inflam_infiltration': {'description': 'Numeric value to represent local response to cellular injury, marked by capillary dilatation, edema and leukocyte infiltration; clinically, inflammation is manifest by reddness, heat, pain, swelling and loss of function, with the need to heal damaged tissue.\\n',\n",
       "   'termDef': {'cde_id': 2897695,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Specimen Inflammation Change Percentage Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2897695&version=1.0'}},\n",
       "  'percent_lymphocyte_infiltration': {'description': 'Numeric value to represent the percentage of infiltration by lymphocytes in a solid tissue normal sample or specimen.\\n',\n",
       "   'termDef': {'cde_id': 2897710,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Specimen Lymphocyte Infiltration Percentage Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2897710&version=2.0'}},\n",
       "  'percent_monocyte_infiltration': {'description': 'Numeric value to represent the percentage of monocyte infiltration in a sample or specimen.\\n',\n",
       "   'termDef': {'cde_id': 5455535,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Specimen Monocyte Infiltration Percentage Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5455535&version=1.0'}},\n",
       "  'percent_necrosis': {'description': 'Numeric value to represent the percentage of cell death in a malignant tumor sample or specimen.\\n',\n",
       "   'termDef': {'cde_id': 2841237,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Malignant Neoplasm Necrosis Percentage Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2841237&version=1.0'}},\n",
       "  'percent_neutrophil_infiltration': {'description': 'Numeric value to represent the percentage of infiltration by neutrophils in a tumor sample or specimen.\\n',\n",
       "   'termDef': {'cde_id': 2841267,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Malignant Neoplasm Neutrophil Infiltration Percentage Cell Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2841267&version=1.0'}},\n",
       "  'percent_normal_cells': {'description': 'Numeric value to represent the percentage of normal cell content in a malignant tumor sample or specimen.\\n',\n",
       "   'termDef': {'cde_id': 2841233,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Malignant Neoplasm Normal Cell Percentage Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2841233&version=1.0'}},\n",
       "  'percent_stromal_cells': {'description': 'Numeric value to represent the percentage of reactive cells that are present in a malignant tumor sample or specimen but are not malignant such as fibroblasts, vascular structures, etc.\\n',\n",
       "   'termDef': {'cde_id': 2841241,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Malignant Neoplasm Stromal Cell Percentage Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2841241&version=1.0'}},\n",
       "  'percent_tumor_cells': {'description': 'Numeric value that represents the percentage of infiltration by granulocytes in a sample.\\n',\n",
       "   'termDef': {'cde_id': 5432686,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Specimen Tumor Cell Percentage Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432686&version=1.0'}},\n",
       "  'percent_tumor_nuclei': {'description': 'Numeric value to represent the percentage of tumor nuclei in a malignant neoplasm sample or specimen.\\n',\n",
       "   'termDef': {'cde_id': 2841225,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Malignant Neoplasm Neoplasm Nucleus Percentage Cell Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2841225&version=1.0'}},\n",
       "  'perineural_invasion_present': {'description': 'a yes/no indicator to ask if perineural invasion or infiltration of tumor or cancer is present.\\n',\n",
       "   'termDef': {'cde_id': 64181,\n",
       "    'cde_version': 3,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tumor Perineural Invasion Ind',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=64181&version=3.0'}},\n",
       "  'phenotype_hpo_id': {'description': 'The ID of the term from Human Phenotype Ontology which represents a harmonized phenotype. (CMG)\\n',\n",
       "   'termDef': {'cde_id': 'HPO',\n",
       "    'cde_version': 1,\n",
       "    'source': 'HPO',\n",
       "    'term': 'The ID of the term from Human Phenotype Ontology which represents a harmonized phenotype. (CMG)',\n",
       "    'term_url': 'https://hpo.jax.org/app/'}},\n",
       "  'platform': {'description': 'Name of the platform used to obtain data.\\n'},\n",
       "  'portion_number': {'description': 'Numeric value that represents the sequential number assigned to a portion of the sample.\\n',\n",
       "   'termDef': {'cde_id': 5432711,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Biospecimen Portion Sequence Number',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432711&version=1.0'}},\n",
       "  'portion_weight': {'description': 'Numeric value that represents the sample portion weight, measured in milligrams.\\n',\n",
       "   'termDef': {'cde_id': 5432593,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Biospecimen Portion Weight Milligram Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432593&version=1.0'}},\n",
       "  'preservation_method': {'description': 'Text term that represents the method used to preserve the sample.\\n',\n",
       "   'termDef': {'cde_id': 5432521,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tissue Sample Preservation Method Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432521&version=1.0'}},\n",
       "  'prior_malignancy': {'description': \"Text term to describe the patient's history of prior cancer diagnosis and the spatial location of any previous cancer occurrence.\\n\",\n",
       "   'termDef': {'cde_id': 3382736,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Prior Cancer Diagnosis Occurrence Description Text',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3382736&version=2.0'}},\n",
       "  'prior_treatment': {'description': 'A yes/no/unknown/not applicable indicator related to the administration of therapeutic agents received before the body specimen was collected.\\n',\n",
       "   'termDef': {'cde_id': 4231463,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Therapeutic Procedure Prior Specimen Collection Administered Yes No Unknown Not Applicable Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4231463&version=1.0'}},\n",
       "  'progesterone_receptor_percent_positive_ihc': {'description': 'Classification to represent Progesterone Receptor Positive results expressed as a percentage value.\\n',\n",
       "   'termDef': {'cde_id': 3128342,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Progesterone Receptor Level Cell Percentage Category',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3128342&version=1.0'}},\n",
       "  'progesterone_receptor_result_ihc': {'description': 'Text term to represent the overall result of Progresterone Receptor (PR) testing.\\n',\n",
       "   'termDef': {'cde_id': 2957357,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Breast Carcinoma Progesterone Receptor Status',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2957357&version=2.0'}},\n",
       "  'progression_or_recurrence': {'description': 'Yes/No/Unknown indicator to identify whether a patient has had a new tumor event after initial treatment.\\n',\n",
       "   'termDef': {'cde_id': 3121376,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'New Neoplasm Event Post Initial Therapy Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3121376&version=1.0'}},\n",
       "  'project_id': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'},\n",
       "  'qc_metric_state': {'description': 'State classification given by FASTQC for the metric. Metric specific details about the states are available on their website.\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': 'FastQC',\n",
       "    'term': 'QC Metric State',\n",
       "    'term_url': 'http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/'}},\n",
       "  'race': {'description': 'An arbitrary classification of a taxonomic group that is a division of a species (HARMONIZED). It usually arises as a consequence of geographical isolation within a species and is characterized by shared heredity, physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau. (GTEx)\\n',\n",
       "   'termDef': {'cde_id': 2192199,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Race Category Text',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2192199&version=1.0'}},\n",
       "  'read_group_name': {'description': 'The name of the read group.\\n'},\n",
       "  'read_length': {'description': 'The length of the reads.\\n'},\n",
       "  'relationship_age_at_diagnosis': {'description': \"The age (in years) when the patient's relative was first diagnosed.\\n\",\n",
       "   'termDef': {'cde_id': 5300571,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Relative Diagnosis Age Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5300571&version=1.0'}},\n",
       "  'relationship_to_proband': {'description': 'The subgroup that describes the state of connectedness between members of the unit of society organized around kinship ties.\\n',\n",
       "   'termDef': {'cde_id': 2690165,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Family Member Relationship Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2690165&version=2.0'}},\n",
       "  'relationship_type': {'description': 'The subgroup that describes the state of connectedness between members of the unit of society organized around kinship ties.\\n',\n",
       "   'termDef': {'cde_id': 2690165,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Family Member Relationship Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2690165&version=2.0'}},\n",
       "  'relative_with_cancer_history': {'description': \"Indicator to signify whether or not an individual's biological relative has been diagnosed with another type of cancer.\\n\",\n",
       "   'termDef': {'cde_id': 3901752,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Other Cancer Biological Relative History Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3901752&version=1.0'}},\n",
       "  'residual_disease': {'description': 'Text terms to describe the status of a tissue margin following surgical resection.\\n',\n",
       "   'termDef': {'cde_id': 2608702,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Surgical Margin Resection Status',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2608702&version=1.0'}},\n",
       "  'sample_type': {'description': 'Text term to describe the source of a biospecimen used for a laboratory test.\\n',\n",
       "   'termDef': {'cde_id': 3111302,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Specimen Type Collection Biospecimen Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3111302&version=2.0'}},\n",
       "  'sample_type_id': {'description': 'The accompanying sample type id for the sample type.\\n'},\n",
       "  'section_location': {'description': 'Tissue source of the slide.\\n'},\n",
       "  'sequencing_center': {'description': 'Name of the center that provided the sequence files.\\n'},\n",
       "  'sex': {'description': 'Text designations that identify gender (HARMONIZED). Gender is described as the assemblage of properties that distinguish people on the basis of their societal roles. [Explanatory Comment 1: Identification of gender is based upon self-report and may come from a form, questionnaire, interview, etc.] (CMG)\\n',\n",
       "   'termDef': {'cde_id': 2200604,\n",
       "    'cde_version': 3,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Person Gender Text Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2200604&version=3.0'}},\n",
       "  'shortest_dimension': {'description': 'Numeric value that represents the shortest dimension of the sample, measured in millimeters.\\n',\n",
       "   'termDef': {'cde_id': 5432603,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tissue Sample Short Dimension Millimeter Measurement',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432603&version=1.0'}},\n",
       "  'site_of_resection_or_biopsy': {'description': 'The third edition of the International Classification of Diseases for Oncology, published in 2000, used principally in tumor and cancer registries for coding the site (topography) and the histology (morphology) of neoplasms. The description of an anatomical region or of a body part. Named locations of, or within, the body. A system of numbered categories for representation of data.\\n',\n",
       "   'termDef': {'cde_id': 3226281,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'International Classification of Diseases for Oncology, Third Edition ICD-O-3 Site Code',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3226281&version=1.0'}},\n",
       "  'size_selection_range': {'description': 'Range of size selection.\\n'},\n",
       "  'smoking_history': {'description': 'Category describing current smoking status and smoking history as self-reported by a patient.\\n',\n",
       "   'termDef': {'cde_id': 2181650,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Smoking History',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2181650&version=1.0'}},\n",
       "  'smoking_intensity': {'description': 'Numeric computed value to represent lifetime tobacco exposure defined as number of cigarettes smoked per day x number of years smoked divided by 20\\n',\n",
       "   'termDef': {'cde_id': 2955385,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Person Cigarette Smoking History Pack Year Value',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2955385&version=1.0'}},\n",
       "  'source_center': {'description': 'Name of the center that provided the item.\\n'},\n",
       "  'spectrophotometer_method': {'description': 'Name of the method used to determine the concentration of purified nucleic acid within a solution.\\n',\n",
       "   'termDef': {'cde_id': 3008378,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Purification Nucleic Acid Solution Concentration Determination Method Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3008378&version=1.0'}},\n",
       "  'spike_ins_concentration': {'description': 'Spike in concentration.\\n'},\n",
       "  'spike_ins_fasta': {'description': 'Name of the FASTA file that contains the spike-in sequences.\\n'},\n",
       "  'state': {'description': 'The current state of the object.\\n'},\n",
       "  'target_capture_kit_catalog_number': {'description': 'Catalog of Target Capture Kit.\\n'},\n",
       "  'target_capture_kit_name': {'description': 'Name of Target Capture Kit.\\n'},\n",
       "  'target_capture_kit_target_region': {'description': 'Target Capture Kit BED file.\\n'},\n",
       "  'target_capture_kit_vendor': {'description': 'Vendor of Target Capture Kit.\\n'},\n",
       "  'target_capture_kit_version': {'description': 'Version of Target Capture Kit.\\n'},\n",
       "  'therapeutic_agents': {'description': 'Text identification of the individual agent(s) used as part of a prior treatment regimen.\\n',\n",
       "   'termDef': {'cde_id': 2975232,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Prior Therapy Regimen Text',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2975232&version=1.0'}},\n",
       "  'time_between_clamping_and_freezing': {'description': 'Numeric representation of the elapsed time between the surgical clamping of blood supply and freezing of the sample, measured in minutes.\\n',\n",
       "   'termDef': {'cde_id': 5432611,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tissue Sample Clamping and Freezing Elapsed Minute Time',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432611&version=1.0'}},\n",
       "  'time_between_excision_and_freezing': {'description': 'Numeric representation of the elapsed time between the excision and freezing of the sample, measured in minutes.\\n',\n",
       "   'termDef': {'cde_id': 5432612,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tissue Sample Excision and Freezing Elapsed Minute Time',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432612&version=1.0'}},\n",
       "  'tissue_or_organ_of_origin': {'description': 'Text term that describes the anatomic site of the tumor or disease.\\n',\n",
       "   'termDef': {'cde_id': 3427536,\n",
       "    'cde_version': 3,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tumor Disease Anatomic Site',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3427536&version=3.0'}},\n",
       "  'tissue_type': {'description': 'Text term that represents a description of the kind of tissue collected with respect to disease status or proximity to tumor tissue.\\n',\n",
       "   'termDef': {'cde_id': 5432687,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tissue Sample Description Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432687&version=1.0'}},\n",
       "  'to_trim_adapter_sequence': {'description': 'Does the user suggest adapter trimming?\\n'},\n",
       "  'tobacco_smoking_onset_year': {'description': 'The year in which the participant began smoking.\\n',\n",
       "   'termDef': {'cde_id': 2228604,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Started Smoking Year',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2228604&version=1.0'}},\n",
       "  'tobacco_smoking_quit_year': {'description': 'The year in which the participant quit smoking.\\n',\n",
       "   'termDef': {'cde_id': 2228610,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Stopped Smoking Year',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2228610&version=1.0'}},\n",
       "  'tobacco_smoking_status': {'description': 'Category describing current smoking status and smoking history as self-reported by a patient.\\n',\n",
       "   'termDef': {'cde_id': 2181650,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Patient Smoking History Category',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2181650&version=1.0'}},\n",
       "  'total_sequences': {'description': 'A count of the total number of sequences processed.\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': 'FastQC',\n",
       "    'term': 'Total Sequences',\n",
       "    'term_url': 'http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/1%20Basic%20Statistics.html'}},\n",
       "  'treatment_anatomic_site': {'description': 'The anatomic site or field targeted by a treatment regimen or single agent therapy.\\n',\n",
       "   'termDef': {'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'source': None,\n",
       "    'term': 'Treatment Anatomic Site',\n",
       "    'term_url': None}},\n",
       "  'treatment_intent_type': {'description': 'Text term to identify the reason for the administration of a treatment regimen. [Manually-curated]\\n',\n",
       "   'termDef': {'cde_id': 2793511,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Treatment Regimen Intent Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2793511&version=1.0'}},\n",
       "  'treatment_or_therapy': {'description': 'A yes/no/unknown/not applicable indicator related to the administration of therapeutic agents received before the body specimen was collected.\\n',\n",
       "   'termDef': {'cde_id': 4231463,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Therapeutic Procedure Prior Specimen Collection Administered Yes No Unknown Not Applicable Indicator',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4231463&version=1.0'}},\n",
       "  'treatment_outcome': {'description': 'Text term that describes the patientÂ¿s final outcome after the treatment was administered.\\n',\n",
       "   'termDef': {'cde_id': 5102383,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Treatment Outcome Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5102383&version=1.0'}},\n",
       "  'treatment_type': {'description': 'Text term that describes the kind of treatment administered.\\n',\n",
       "   'termDef': {'cde_id': 5102381,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Treatment Method Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5102381&version=1.0'}},\n",
       "  'tumor_code': {'description': 'Diagnostic tumor code of the tissue sample source.\\n'},\n",
       "  'tumor_code_id': {'description': 'BCR-defined id code for the tumor sample.\\n'},\n",
       "  'tumor_descriptor': {'description': 'Text that describes the kind of disease present in the tumor specimen as related to a specific timepoint.\\n',\n",
       "   'termDef': {'cde_id': 3288124,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tumor Tissue Disease Description Type',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3288124&version=1.0'}},\n",
       "  'tumor_grade': {'description': 'Numeric value to express the degree of abnormality of cancer cells, a measure of differentiation and aggressiveness.\\n',\n",
       "   'termDef': {'cde_id': 2785839,\n",
       "    'cde_version': 2,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Neoplasm Histologic Grade',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2785839&version=2.0'}},\n",
       "  'tumor_stage': {'description': 'The extent of a cancer in the body. Staging is usually based on the size of the tumor, whether lymph nodes contain cancer, and whether the cancer has spread from the original site to other parts of the body. The accepted values for tumor_stage depend on the tumor site, type, and accepted staging system. These items should accompany the tumor_stage value as associated metadata.\\n',\n",
       "   'termDef': {'cde_id': 'C16899',\n",
       "    'cde_version': None,\n",
       "    'source': 'NCIt',\n",
       "    'term': 'Tumor Stage',\n",
       "    'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/pages/concept_details.jsf?dictionary=NCI%20Thesaurus&code=C16899'}},\n",
       "  'vascular_invasion_present': {'description': 'The yes/no indicator to ask if large vessel or venous invasion was detected by surgery or presence in a tumor specimen.\\n',\n",
       "   'termDef': {'cde_id': 64358,\n",
       "    'cde_version': 3,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Tumor Vascular Invasion Ind-3',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=64358&version=3.0'}},\n",
       "  'vital_status': {'description': 'The survival state of the person registered on the protocol.\\n',\n",
       "   'termDef': {'cde_id': 5,\n",
       "    'cde_version': 5,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Patient Vital Status',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5&version=5.0'}},\n",
       "  'weight': {'description': 'The weight of the patient measured in lbs (HARMONIZED).\\n',\n",
       "   'termDef': {'cde_id': 651,\n",
       "    'cde_version': 4,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Patient Weight Measurement',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=651&version=4.0'}},\n",
       "  'well_number': {'description': 'Numeric value that represents the the well location within a plate for the analyte or aliquot from the sample.\\n',\n",
       "   'termDef': {'cde_id': 5432613,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Biospecimen Analyte or Aliquot Plate Well Number',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432613&version=1.0'}},\n",
       "  'workflow_type': {'description': 'Generic name for the workflow used to analyze a data set.\\n'},\n",
       "  'year_of_birth': {'description': 'Numeric value to represent the calendar year in which an individual was born.\\n',\n",
       "   'termDef': {'cde_id': 2896954,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Year Birth Date Number',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2896954&version=1.0'}},\n",
       "  'year_of_death': {'description': 'Numeric value to represent the year of the death of an individual.\\n',\n",
       "   'termDef': {'cde_id': 2897030,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Year Death Number',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2897030&version=1.0'}},\n",
       "  'year_of_diagnosis': {'description': \"Numeric value to represent the year of an individual's initial pathologic diagnosis of cancer.\\n\",\n",
       "   'termDef': {'cde_id': 2896960,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Year of initial pathologic diagnosis',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2896960&version=1.0'}},\n",
       "  'years_smoked': {'description': \"Numeric value (or unknown) to represent the number of years a person has been smoking (HARMONIZED). If the number of years is greater than 89, see 'years_smoked_gt89'.\\n\",\n",
       "   'termDef': {'cde_id': 3137957,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Person Smoking Duration Year Count',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3137957&version=1.0'}},\n",
       "  'years_smoked_gt89': {'description': 'Indicate whether the numeric value to represent the number of years a person has been smoking (HARMONIZED) is greater than 89 years.\\n',\n",
       "   'termDef': {'cde_id': 3137957,\n",
       "    'cde_version': 1,\n",
       "    'source': 'caDSR',\n",
       "    'term': 'Person Smoking Duration Year Count',\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3137957&version=1.0'}}},\n",
       " 'aligned_reads_index_file.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'data_file',\n",
       "  'description': 'Data file containing an index for a set of aligned reads',\n",
       "  'id': 'aligned_reads_index_file',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'aligned_reads_index_files',\n",
       "      'label': 'describes',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'aligned_reads_files',\n",
       "      'required': False,\n",
       "      'target_type': 'aligned_reads_file'},\n",
       "     {'backref': 'aligned_reads_index_files',\n",
       "      'label': 'generated_from',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'alignment_workflows',\n",
       "      'required': False,\n",
       "      'target_type': 'alignment_workflow'},\n",
       "     {'backref': 'aligned_reads_index_files',\n",
       "      'label': 'describes',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'core_metadata_collections',\n",
       "      'required': False,\n",
       "      'target_type': 'core_metadata_collection'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/data_file_properties',\n",
       "   'aligned_reads_files': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'alignment_workflows': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'core_metadata_collections': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'data_category': {'description': 'Broad categorization of the contents of the data file.',\n",
       "    'enum': ['analysis',\n",
       "     'sequencing reads',\n",
       "     'single nucleotide variation',\n",
       "     'transcriptome profiling',\n",
       "     'supplemental']},\n",
       "   'data_format': {'description': 'Format of the data files.',\n",
       "    'enum': ['bai', 'bam.bai', 'crai', 'cram.crai']},\n",
       "   'data_type': {'description': 'Specific content type of the data file.',\n",
       "    'enum': ['unaligned reads',\n",
       "     'aligned reads',\n",
       "     'variants annotation',\n",
       "     'clinical supplement']}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'data_category',\n",
       "   'data_format',\n",
       "   'data_type'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Aligned Reads Index File',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'variant_workflow.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'analysis',\n",
       "  'description': 'Metadata for the variant calling pipeline used to call germline variants',\n",
       "  'id': 'variant_workflow',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'variant_workflows',\n",
       "      'label': 'part_of',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'aligned_reads_files',\n",
       "      'required': False,\n",
       "      'target_type': 'aligned_reads_file'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'aligned_reads_files': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'workflow_end_datetime': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]',\n",
       "    'pattern': '^\\\\d{4}-\\\\d{2}-\\\\d{2}T\\\\d{2}:\\\\d{2}:\\\\d{2}(?:\\\\.\\\\d+)?(?:Z|[+-]\\\\d{2}:\\\\d{2})$',\n",
       "    'type': 'string'},\n",
       "   'workflow_link': {'description': 'Link to Github hash for the CWL workflow used.',\n",
       "    'type': 'string'},\n",
       "   'workflow_start_datetime': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]',\n",
       "    'pattern': '^\\\\d{4}-\\\\d{2}-\\\\d{2}T\\\\d{2}:\\\\d{2}:\\\\d{2}(?:\\\\.\\\\d+)?(?:Z|[+-]\\\\d{2}:\\\\d{2})$',\n",
       "    'type': 'string'},\n",
       "   'workflow_type': {'description': 'Generic name for the workflow used to analyze a data set',\n",
       "    'enum': ['HaplotypeCaller',\n",
       "     'DeepVariant',\n",
       "     'Strelka2',\n",
       "     'FreeBayes',\n",
       "     'VarScan']},\n",
       "   'workflow_version': {'description': 'Version of the workflow used',\n",
       "    'type': 'string'}},\n",
       "  'required': ['type', 'submitter_id', 'workflow_type'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Variant Workflow',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'proteomics_file.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'data_file',\n",
       "  'description': 'Data file that contains the results of a proteomics assay',\n",
       "  'id': 'proteomics_file',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'proteomics_files',\n",
       "      'label': 'derives_from',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'proteomics_assays',\n",
       "      'required': False,\n",
       "      'target_type': 'proteomics_assay'},\n",
       "     {'backref': 'proteomics_files',\n",
       "      'label': 'describes',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'core_metadata_collections',\n",
       "      'required': False,\n",
       "      'target_type': 'core_metadata_collection'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/data_file_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'core_metadata_collections': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'cv': {'description': 'Coefficient of variation (%)',\n",
       "    'type': ['number', 'null']},\n",
       "   'data_category': {'description': 'Broad categorization of the contents of the data file.',\n",
       "    'enum': ['mass spec raw', 'mass spec analysed', 'summarised results']},\n",
       "   'data_format': {'description': 'The format of the data in this data file',\n",
       "    'enum': ['mgf']},\n",
       "   'data_type': {'description': 'The type of data in this data file',\n",
       "    'enum': ['MS', 'MS/MS']},\n",
       "   'proteomic_unit': {'description': 'Units of measurement captured in the proteomics file.',\n",
       "    'enum': ['pmol/mL', 'umol/mL']},\n",
       "   'proteomics_assays': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'data_category',\n",
       "   'data_format',\n",
       "   'data_type'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Proteomics File',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'exposure.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'clinical',\n",
       "  'description': 'Clinically relevant patient information relating to environmental exposures (e.g. smoking)',\n",
       "  'id': 'exposure',\n",
       "  'links': [{'backref': 'exposures',\n",
       "    'label': 'describes',\n",
       "    'multiplicity': 'one_to_one',\n",
       "    'name': 'subjects',\n",
       "    'required': True,\n",
       "    'target_type': 'subject'}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'cigarettes_per_day': {'description': 'Self reported cigarettes per day.',\n",
       "    'type': 'integer'},\n",
       "   'drinking_status': {'description': 'Self reported current drinking status.',\n",
       "    'enum': ['never', 'former', 'current']},\n",
       "   'drinks_per_week': {'description': 'Estimated standard drinks per week',\n",
       "    'type': 'integer'},\n",
       "   'smoking_status': {'description': 'Indicates whether subject ever regularly smoked cigarettes.',\n",
       "    'enum': ['never', 'former', 'current']},\n",
       "   'subjects': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Exposure',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'metabolomics_assay.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'experimental_methods',\n",
       "  'description': 'Details about the methods that produced the metabolomics assay',\n",
       "  'id': 'metabolomics_assay',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'metabolomics_assays',\n",
       "      'label': 'produces',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'samples',\n",
       "      'required': False,\n",
       "      'target_type': 'sample'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'assay_description': {'description': 'A description of the type of assay used.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'metabolite_names': {'description': 'A list of all metabolites assayed in this method',\n",
       "    'items': {'enum': ['1-methylhistamine',\n",
       "      '2-Arachidonyl glycerol',\n",
       "      \"2'-deoxyadenosine\",\n",
       "      '3-deaazadenosine',\n",
       "      '3HK',\n",
       "      '3-IPA',\n",
       "      '5-Aminolevulinic Acid',\n",
       "      'Acetylcarnitine',\n",
       "      'Adenosine',\n",
       "      'Alanine',\n",
       "      'Anandamide',\n",
       "      'Arachidonic acid',\n",
       "      'Arginine',\n",
       "      'Asparagine',\n",
       "      'Aspartate',\n",
       "      'Betaine',\n",
       "      'Butyrylcarnitine',\n",
       "      'cAMP',\n",
       "      'Carnitine',\n",
       "      'Choline',\n",
       "      'Colchicine',\n",
       "      'Creatine',\n",
       "      'Cysteamine',\n",
       "      'Cysteine',\n",
       "      'Cytosine',\n",
       "      'DMGV',\n",
       "      'GlucosePos2',\n",
       "      'Glutamate',\n",
       "      'Glutamine',\n",
       "      'Histidine',\n",
       "      'Isoleucine_Leucine',\n",
       "      'Kynurenic acid',\n",
       "      'L-Homoserine',\n",
       "      'Methionine',\n",
       "      'Phenylalanine',\n",
       "      'Phosphocholine',\n",
       "      'Proline',\n",
       "      'Riboflavin',\n",
       "      'Serine',\n",
       "      'Serotonin',\n",
       "      'Spermine',\n",
       "      'Taurine',\n",
       "      'Thiamine',\n",
       "      'Threonine',\n",
       "      'Thymidine',\n",
       "      'TMNO',\n",
       "      'trans-HYP',\n",
       "      'Tyrosine',\n",
       "      'Uridine',\n",
       "      'Valine',\n",
       "      'Valine-d8',\n",
       "      'Î±-keto-Î²-methylvaleric acid.1',\n",
       "      'Î±-keto-Î²-methylvaleric acid.2'],\n",
       "     'enumDef': [{'enumeration': '1-methylhistamine',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0000898'},\n",
       "      {'enumeration': '2-Arachidonyl glycerol',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0004666'},\n",
       "      {'enumeration': \"2'-deoxyadenosine\",\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0000101'},\n",
       "      {'enumeration': '3-deaazadenosine',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0029956'},\n",
       "      {'enumeration': '3HK', 'source': 'HMDB', 'term_id': 'HMDB0011631'},\n",
       "      {'enumeration': '3-IPA', 'source': 'HMDB', 'term_id': 'HMDB0002302'},\n",
       "      {'enumeration': '5-Aminolevulinic Acid',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0001149'},\n",
       "      {'enumeration': 'Acetylcarnitine',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0000201'},\n",
       "      {'enumeration': 'Adenosine', 'source': 'HMDB', 'term_id': 'HMDB0000050'},\n",
       "      {'enumeration': 'Alanine', 'source': 'HMDB', 'term_id': 'HMDB0000161'},\n",
       "      {'enumeration': 'Anandamide',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0004080'},\n",
       "      {'enumeration': 'Arachidonic acid',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0001043'},\n",
       "      {'enumeration': 'Arginine', 'source': 'HMDB', 'term_id': 'HMDB0000517'},\n",
       "      {'enumeration': 'Asparagine',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0000168'},\n",
       "      {'enumeration': 'Aspartate', 'source': 'HMDB', 'term_id': 'HMDB0000191'},\n",
       "      {'enumeration': 'Betaine', 'source': 'HMDB', 'term_id': 'HMDB0000043'},\n",
       "      {'enumeration': 'Butyrylcarnitine',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0002013'},\n",
       "      {'enumeration': 'cAMP', 'source': 'HMDB', 'term_id': 'HMDB0000058'},\n",
       "      {'enumeration': 'Carnitine', 'source': 'HMDB', 'term_id': 'HMDB0000062'},\n",
       "      {'enumeration': 'Choline', 'source': 'HMDB', 'term_id': 'HMDB0000097'},\n",
       "      {'enumeration': 'Colchicine',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0015466'},\n",
       "      {'enumeration': 'Creatine', 'source': 'HMDB', 'term_id': 'HMDB0000064'},\n",
       "      {'enumeration': 'Cysteamine',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0002991'},\n",
       "      {'enumeration': 'Cysteine', 'source': 'HMDB', 'term_id': 'HMDB0000574'},\n",
       "      {'enumeration': 'Cytosine', 'source': 'HMDB', 'term_id': 'HMDB0000630'},\n",
       "      {'enumeration': 'DMGV', 'source': 'HMDB', 'term_id': 'HMDB0240212'},\n",
       "      {'enumeration': 'GlucosePos2',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0000122'},\n",
       "      {'enumeration': 'Glutamate', 'source': 'HMDB', 'term_id': 'HMDB0000148'},\n",
       "      {'enumeration': 'Glutamine', 'source': 'HMDB', 'term_id': 'HMDB0000641'},\n",
       "      {'enumeration': 'Histidine', 'source': 'HMDB', 'term_id': 'HMDB0000177'},\n",
       "      {'enumeration': 'Isoleucine_Leucine',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0028911'},\n",
       "      {'enumeration': 'Kynurenic acid',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0000715'},\n",
       "      {'enumeration': 'L-Homoserine',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0000719'},\n",
       "      {'enumeration': 'Methionine',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0000696'},\n",
       "      {'enumeration': 'Phenylalanine',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0000159'},\n",
       "      {'enumeration': 'Phosphocholine',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0001565'},\n",
       "      {'enumeration': 'Proline', 'source': 'HMDB', 'term_id': 'HMDB0000162'},\n",
       "      {'enumeration': 'Riboflavin',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0000244'},\n",
       "      {'enumeration': 'Serine', 'source': 'HMDB', 'term_id': 'HMDB0062263'},\n",
       "      {'enumeration': 'Serotonin', 'source': 'HMDB', 'term_id': 'HMDB0000259'},\n",
       "      {'enumeration': 'Spermine', 'source': 'HMDB', 'term_id': 'HMDB0001256'},\n",
       "      {'enumeration': 'Taurine', 'source': 'HMDB', 'term_id': 'HMDB0000251'},\n",
       "      {'enumeration': 'Thiamine', 'source': 'HMDB', 'term_id': 'HMDB0000235'},\n",
       "      {'enumeration': 'Threonine', 'source': 'HMDB', 'term_id': 'HMDB0000167'},\n",
       "      {'enumeration': 'Thymidine', 'source': 'HMDB', 'term_id': 'HMDB0000273'},\n",
       "      {'enumeration': 'TMNO', 'source': 'HMDB', 'term_id': 'HMDB0000925'},\n",
       "      {'enumeration': 'trans-HYP', 'source': 'HMDB', 'term_id': 'HMDB0000725'},\n",
       "      {'enumeration': 'Tyrosine', 'source': 'HMDB', 'term_id': 'HMDB0000158'},\n",
       "      {'enumeration': 'Uridine', 'source': 'HMDB', 'term_id': 'HMDB0000296'},\n",
       "      {'enumeration': 'Valine', 'source': 'HMDB', 'term_id': 'HMDB0000883'},\n",
       "      {'enumeration': 'Valine-d8', 'source': 'HMDB', 'term_id': 'HMDB0000883'},\n",
       "      {'enumeration': 'Î±-keto-Î²-methylvaleric acid.1',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0000491'},\n",
       "      {'enumeration': 'Î±-keto-Î²-methylvaleric acid.2',\n",
       "       'source': 'HMDB',\n",
       "       'term_id': 'HMDB0000491'}]},\n",
       "    'type': 'array'},\n",
       "   'samples': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type', 'submitter_id', 'assay_description'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Metabolomics Assay',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'lipidomics_mapping_file.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'data_file',\n",
       "  'description': 'Data file containing the lipid species names and lipid classes, along with their mapping to \"LIPID MAPS\"',\n",
       "  'id': 'lipidomics_mapping_file',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'lipidomics_mapping_files',\n",
       "      'label': 'part_of',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'lipidomics_files',\n",
       "      'required': False,\n",
       "      'target_type': 'lipidomics_file'},\n",
       "     {'backref': 'lipidomics_mapping_files',\n",
       "      'label': 'describes',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'core_metadata_collections',\n",
       "      'required': False,\n",
       "      'target_type': 'core_metadata_collection'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/data_file_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'core_metadata_collections': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'data_category': {'description': 'Broad categorization of the contents of the data file.',\n",
       "    'enum': ['mass spec raw', 'mass spec analysed', 'summarised results']},\n",
       "   'data_format': {'description': 'The format of the data in this data file',\n",
       "    'enum': ['csv', 'tsv', 'txt']},\n",
       "   'data_type': {'description': 'Specific content type of the data file.',\n",
       "    'enum': ['unaligned reads',\n",
       "     'aligned reads',\n",
       "     'variants annotation',\n",
       "     'clinical supplement']},\n",
       "   'lipidomics_files': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'mapping_file_version': {'description': 'Version of the mapping file',\n",
       "    'type': 'string'}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'data_category',\n",
       "   'data_format',\n",
       "   'data_type',\n",
       "   'mapping_file_version'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Lipidomics Naming Mapping File',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'lipidomics_file.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'data_file',\n",
       "  'description': 'Data file containing lipidomics data',\n",
       "  'id': 'lipidomics_file',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'lipidomics_files',\n",
       "      'label': 'data_from',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'lipidomics_assays',\n",
       "      'required': False,\n",
       "      'target_type': 'lipidomics_assay'},\n",
       "     {'backref': 'lipidomics_files',\n",
       "      'label': 'data_from',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'core_metadata_collections',\n",
       "      'required': False,\n",
       "      'target_type': 'core_metadata_collection'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/data_file_properties',\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'core_metadata_collections': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'cv': {'description': 'Coefficient of variation (%)',\n",
       "    'type': ['number', 'null']},\n",
       "   'data_category': {'description': 'Broad categorization of the contents of the data file.',\n",
       "    'enum': ['mass spec raw', 'mass spec analysed', 'summarised results']},\n",
       "   'data_format': {'description': 'The format of the data in this data file',\n",
       "    'enum': ['csv', 'tsv', 'txt']},\n",
       "   'data_type': {'description': 'The type of data in this data file',\n",
       "    'enum': ['MS', 'MS/MS']},\n",
       "   'imputed_data': {'description': 'Does the lipidomic file contain imputed data for missing lipid species?',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'individual_or_sum': {'description': 'Does the lipidomic file contain data for individual lipid species? Or total lipid class?',\n",
       "    'enum': ['individual', 'sum']},\n",
       "   'lipidomic_unit': {'description': 'Units of measurement captured in the lipidomics file.',\n",
       "    'enum': ['pmol/mL', 'umol/mL']},\n",
       "   'lipidomics_assays': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'data_category',\n",
       "   'data_format',\n",
       "   'data_type',\n",
       "   'imputed_data',\n",
       "   'individual_or_sum'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Lipidomics File',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'aligned_reads_file.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'data_file',\n",
       "  'description': 'Data file containing aligned reads from a sequencing experiment.',\n",
       "  'id': 'aligned_reads_file',\n",
       "  'links': [{'exclusive': None,\n",
       "    'required': True,\n",
       "    'subgroup': [{'backref': 'aligned_reads_files',\n",
       "      'label': 'generated_from',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'unaligned_reads_files',\n",
       "      'required': False,\n",
       "      'target_type': 'unaligned_reads_file'},\n",
       "     {'backref': 'aligned_reads_files',\n",
       "      'label': 'generated_from',\n",
       "      'multiplicity': 'many_to_one',\n",
       "      'name': 'alignment_workflows',\n",
       "      'required': False,\n",
       "      'target_type': 'alignment_workflow'},\n",
       "     {'backref': 'aligned_reads_files',\n",
       "      'label': 'data_from',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'name': 'core_metadata_collections',\n",
       "      'required': False,\n",
       "      'target_type': 'core_metadata_collection'}]}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/data_file_properties',\n",
       "   'alignment_workflows': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'core_metadata_collections': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'data_category': {'description': 'Broad categorization of the contents of the data file.',\n",
       "    'enum': ['analysis',\n",
       "     'sequencing reads',\n",
       "     'single nucleotide variation',\n",
       "     'transcriptome profiling',\n",
       "     'supplemental']},\n",
       "   'data_format': {'description': 'Format of the data files.',\n",
       "    'enum': ['bam', 'cram']},\n",
       "   'data_type': {'description': 'Specific content type of the data file.',\n",
       "    'enum': ['unaligned reads',\n",
       "     'aligned reads',\n",
       "     'variants annotation',\n",
       "     'clinical supplement']},\n",
       "   'reference_genome_build': {'description': 'Reference genome used e.g. GRCh37.',\n",
       "    'enum': ['GRCh38', 'GRCh37']},\n",
       "   'run_id': {'description': 'Sequencing run ID associated with file',\n",
       "    'type': 'string'},\n",
       "   'unaligned_reads_files': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'data_category',\n",
       "   'data_format',\n",
       "   'data_type'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Aligned Reads File',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'lab_result.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'clinical',\n",
       "  'description': 'Measurements obtained from blood or other laboratory tests required by a medical professional.',\n",
       "  'id': 'lab_result',\n",
       "  'links': [{'backref': 'lab_results',\n",
       "    'label': 'describes',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'name': 'subjects',\n",
       "    'required': True,\n",
       "    'target_type': 'subject'}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'age_at_collection': {'description': 'Age of the participant at the time of lab tests.',\n",
       "    'type': ['number', 'null']},\n",
       "   'alternate_timepoint': {'description': 'If the data is not a baseline measurement, the timepoint name is defined here.',\n",
       "    'type': 'string'},\n",
       "   'baseline_timepoint': {'description': 'Does the data reflect a baseline measurement?',\n",
       "    'type': 'boolean'},\n",
       "   'bnp': {'description': 'Brain naturemic peptide (BNP) in ng/L',\n",
       "    'type': ['number', 'null']},\n",
       "   'cac_score': {'description': 'Coronary artery calcification (CAC). Score using Agatston scoring of CT scan(s) of coronary arteries',\n",
       "    'type': ['number', 'null']},\n",
       "   'creatinine_serum_enzymatic': {'description': 'Serum creatinine level measured by enzymatic assay (Âµmol/L)',\n",
       "    'type': ['number', 'null']},\n",
       "   'creatinine_serum_jaffe': {'description': 'Serum creatinine level measured by jaffe assay (Âµmol/L)',\n",
       "    'type': ['number', 'null']},\n",
       "   'egfr_baseline_CKDEPI': {'description': 'Estimated glomerular filtration rate (eGFR) in mL/min/1.73m?^2, calcuated using CKD-EPI formula',\n",
       "    'type': ['number', 'null']},\n",
       "   'egfr_baseline_MDRD': {'description': 'Estimated glomerular filtration rate (eGFR) in mL/min/1.73m?^2, calcuated using MDRD formula',\n",
       "    'type': ['number', 'null']},\n",
       "   'fasting': {'description': 'Whether the participant had fasted before blood was collected.',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'glucose_fasting': {'description': 'Fasting blood glucose level (mmol/l) at baseline',\n",
       "    'termDef': [{'source': 'efo',\n",
       "      'term': 'fasting blood glucose measurement',\n",
       "      'term_id': 'EFO:0004465'}],\n",
       "    'type': ['number', 'null']},\n",
       "   'hba1c_ifcc': {'description': 'HbA1c at baseline (IFCC - mmol/mol)',\n",
       "    'termDef': [{'source': 'efo',\n",
       "      'term': 'HbA1c measurement',\n",
       "      'term_id': 'EFO:0004541'}],\n",
       "    'type': ['number', 'null']},\n",
       "   'hba1c_ngsp': {'description': 'HbA1c at baseline (NGSP - %)',\n",
       "    'termDef': [{'source': 'efo',\n",
       "      'term': 'HbA1c measurement',\n",
       "      'term_id': 'EFO:0004541'}],\n",
       "    'type': ['number', 'null']},\n",
       "   'hdl': {'description': 'Blood mass concentration of high-density lipoprotein (HDL-C) cholesterol (mmol/L) at baseline (1 decimal point)',\n",
       "    'termDef': [{'source': 'efo',\n",
       "      'term': 'HDL cholesterol',\n",
       "      'term_id': 'CHEBI:47775'}],\n",
       "    'type': ['number', 'null']},\n",
       "   'ldl': {'description': 'Blood mass concentration of low-density lipoprotein (LDL-C) cholesterol (mmol/L) at baseline (1 decimal point). Note: LDL is not accurate when triglycerides >=4.52 mmol/l',\n",
       "    'termDef': [{'source': 'efo',\n",
       "      'term': 'LDL cholesterol',\n",
       "      'term_id': 'EFO:0004195'}],\n",
       "    'type': ['number', 'null']},\n",
       "   'nt_probnp': {'description': 'NT-proBNP in ng/L',\n",
       "    'type': ['number', 'null']},\n",
       "   'subjects': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'total_cholesterol': {'description': 'Blood mass concentration of total cholesterol (mmol/L) at baseline (1 decimal point)',\n",
       "    'termDef': [{'source': 'efo',\n",
       "      'term': 'cholesterol',\n",
       "      'term_id': 'CHEBI:16113'}],\n",
       "    'type': ['number', 'null']},\n",
       "   'triglycerides': {'description': 'Blood mass concentration of triglycerides (mmol/L) at baseline (1 decimal point)',\n",
       "    'termDef': [{'source': 'efo',\n",
       "      'term': 'triglyceride measurement',\n",
       "      'term_id': 'EFO:0004530'}],\n",
       "    'type': ['number', 'null']},\n",
       "   'uacr': {'description': 'Urine albumin-creatinine ratio (uACR) (measured in mg/mmol).',\n",
       "    'type': ['number', 'null']}},\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'subjects',\n",
       "   'hdl',\n",
       "   'ldl',\n",
       "   'total_cholesterol',\n",
       "   'triglycerides'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Laboratory Result',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'medication.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'clinical',\n",
       "  'description': 'Records about historical or current medication use.',\n",
       "  'id': 'medication',\n",
       "  'links': [{'backref': 'medications',\n",
       "    'label': 'taken_by',\n",
       "    'multiplicity': 'one_to_one',\n",
       "    'name': 'subjects',\n",
       "    'required': True,\n",
       "    'target_type': 'subject'}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'antithrombotic_meds': {'description': 'Self-reported / measured use of antithrobotic medication ',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'bp_lowering_meds': {'description': 'Self-reported use of anti-hypertensive medication at the time of blood pressure measurement.',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'diabetes_therapy': {'description': 'Self-reported use of anti-diabetic therapy.',\n",
       "    'enum': ['diet', 'oral', 'insulin', 'injectible', 'combination', 'other']},\n",
       "   'lipid_lowering_meds': {'description': 'Self-reported use of lipid lowering drugs (are you taking tables to lower your cholesterol?). Full drug audit at wave 2 and 3. ',\n",
       "    'enum': ['yes', 'no']},\n",
       "   'subjects': {'$ref': '_definitions.yaml#/to_one'}},\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Medication Use Record',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'publication.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'administrative',\n",
       "  'description': 'Publication for a project',\n",
       "  'id': 'publication',\n",
       "  'links': [{'backref': 'publications',\n",
       "    'label': 'refers_to',\n",
       "    'multiplicity': 'many_to_many',\n",
       "    'name': 'projects',\n",
       "    'required': True,\n",
       "    'target_type': 'project'}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'doi': {'description': 'Digital object identifier for the published article or pre-print. Starting with 10.',\n",
       "    'pattern': '^10\\\\.\\\\d{4,9}.+',\n",
       "    'type': 'string'},\n",
       "   'is_primary': {'description': 'Whether the publication is the primary publication for this study',\n",
       "    'type': 'boolean'},\n",
       "   'pmid': {'description': 'The pubmed id of the article', 'type': 'string'},\n",
       "   'projects': {'$ref': '_definitions.yaml#/to_many_project'},\n",
       "   'title': {'description': 'Title of the publication', 'type': 'string'},\n",
       "   'url': {'description': 'Direct link to the published article',\n",
       "    'type': 'string'}},\n",
       "  'required': ['type', 'submitter_id', 'projects', 'doi', 'title'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Publication',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'subject.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'clinical',\n",
       "  'description': 'An individual participant in the study with baseline measurements.',\n",
       "  'id': 'subject',\n",
       "  'links': [{'backref': 'subjects',\n",
       "    'label': 'part_of',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'name': 'projects',\n",
       "    'required': True,\n",
       "    'target_type': 'project'}],\n",
       "  'namespace': 'https://data.test.biocommons.org.au/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'cohort_id': {'description': 'Identifier for the cohort the subject is a part of',\n",
       "    'type': 'string'},\n",
       "   'consent_codes': {'description': 'Subject Consent | Data Use Restrictions that are used to indicate  permissions/restrictions for datasets and/or materials, and relates to the purposes for which datasets and/or material might be removed, stored or used. Based on the Data Use Ontology : see http://www.obofoundry.org/ontology/duo.html',\n",
       "    'items': {'enum': ['General Research Use',\n",
       "      'Health or Biomedical Research',\n",
       "      'Disease Specific Research',\n",
       "      'not for profit, non commercial user only',\n",
       "      'ethics approval required',\n",
       "      'user specific restriction'],\n",
       "     'enumDef': [{'enumeration': 'General Research Use',\n",
       "       'source': 'duo',\n",
       "       'term_id': 'DUO:0000042'},\n",
       "      {'enumeration': 'Health or Biomedical Research',\n",
       "       'source': 'duo',\n",
       "       'term_id': 'DUO:0000006'},\n",
       "      {'enumeration': 'Disease Specific Research',\n",
       "       'source': 'duo',\n",
       "       'term_id': 'DUO:0000007'},\n",
       "      {'enumeration': 'not for profit, non commercial user only',\n",
       "       'source': 'duo',\n",
       "       'term_id': 'DUO:0000018'},\n",
       "      {'enumeration': 'ethics approval required',\n",
       "       'source': 'duo',\n",
       "       'term_id': 'DUO:0000021'},\n",
       "      {'enumeration': 'user specific restriction',\n",
       "       'source': 'duo',\n",
       "       'term_id': 'DUO:0000026'}]},\n",
       "    'termDef': [{'term': 'consent_codes'}],\n",
       "    'type': 'array'},\n",
       "   'patient_id': {'description': 'De-identified unique identifier for the patient/subject',\n",
       "    'type': 'string'},\n",
       "   'projects': {'$ref': '_definitions.yaml#/to_one_project'}},\n",
       "  'required': ['type', 'submitter_id', 'projects', 'patient_id'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Subject',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None},\n",
       " 'core_metadata_collection.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'additionalProperties': False,\n",
       "  'category': 'administrative',\n",
       "  'description': 'Structured description of a collection of several dataset\\n',\n",
       "  'id': 'core_metadata_collection',\n",
       "  'links': [{'backref': 'core_metadata_collections',\n",
       "    'label': 'data_from',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'name': 'projects',\n",
       "    'required': True,\n",
       "    'target_type': 'project'}],\n",
       "  'namespace': 'https://dcp.bionimbus.org/',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'contributor': {'description': 'An entity responsible for making contributions to the resource. Examples of a Contributor include a person, an organization, or a service. Typically, the name of a Contributor should be used to indicate the entity.\\n',\n",
       "    'type': 'string'},\n",
       "   'coverage': {'description': 'The spatial or temporal topic of the resource, the spatial applicability of the resource, or the jurisdiction under which the resource is relevant. Spatial topic and spatial applicability may be a named place or a location specified by its geographic coordinates. Temporal topic may be a named period, date, or date range. A jurisdiction may be a named administrative entity or a geographic place to which the resource applies. Recommended best practice is to use a controlled vocabulary such as the Thesaurus of Geographic Names [TGN] (http://www.getty.edu/research/tools/vocabulary/tgn/index.html). Where appropriate, named places or time periods can be used in preference to numeric identifiers such as sets of coordinates or date ranges.\\n',\n",
       "    'type': 'string'},\n",
       "   'creator': {'description': 'An entity primarily responsible for making the resource. Examples of a Creator include a person, an organization, or a service. Typically, the name of a Creator should be used to indicate the entity.\\n',\n",
       "    'type': 'string'},\n",
       "   'data_type': {'description': 'The nature or genre of the resource. Recommended best practice is to use a controlled vocabulary such as the DCMI Type Vocabulary [DCMITYPE]. To describe the file format, physical medium, or dimensions of the resource, use the Format element.\\n',\n",
       "    'type': 'string'},\n",
       "   'date': {'$ref': '_definitions.yaml#/datetime'},\n",
       "   'description': {'description': 'An account of the resource. Description may include but is not limited to: an abstract, a table of contents, a graphical representation, or a free-text account of the resource.\\n',\n",
       "    'type': 'string'},\n",
       "   'format': {'description': 'The file format, physical medium, or dimensions of the resource. Examples of dimensions include size and duration. Recommended best practice is to use a controlled vocabulary such as the list of Internet Media Types [MIME] (http://www.iana.org/assignments/media-types/).\\n',\n",
       "    'type': 'string'},\n",
       "   'language': {'description': 'A language of the resource. Recommended best practice is to use a controlled vocabulary such as RFC 4646 (http://www.ietf.org/rfc/rfc4646.txt).\\n',\n",
       "    'type': 'string'},\n",
       "   'projects': {'$ref': '_definitions.yaml#/to_one_project'},\n",
       "   'publisher': {'description': 'An entity responsible for making the resource available. Examples of a Publisher include a person, an organization, or a service. Typically, the name of a Publisher should be used to indicate the entity.\\n',\n",
       "    'type': 'string'},\n",
       "   'relation': {'description': 'A related resource. Recommended best practice is to identify the related resource by means of a string conforming to a formal identification system.\\xa0\\n',\n",
       "    'type': 'string'},\n",
       "   'rights': {'description': 'Information about rights held in and over the resource. Typically, rights information includes a statement about various property rights associated with the resource, including intellectual property rights.\\n',\n",
       "    'type': 'string'},\n",
       "   'source': {'description': 'A related resource from which the described resource is derived. The described resource may be derived from the related resource in whole or in part. Recommended best practice is to identify the related resource by means of a string conforming to a formal identification system.\\n',\n",
       "    'type': 'string'},\n",
       "   'subject': {'description': 'The topic of the resource. Typically, the subject will be represented using keywords, key phrases, or classification codes. Recommended best practice is to use a controlled vocabulary.\\n',\n",
       "    'type': 'string'},\n",
       "   'title': {'description': 'A name given to the resource. Typically, a Title will be a name by which the resource is formally known.\\n',\n",
       "    'type': 'string'}},\n",
       "  'required': ['submitter_id', 'type', 'projects'],\n",
       "  'submittable': True,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'title': 'Core Metadata Collection',\n",
       "  'type': 'object',\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'validators': None}}"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resolved_schema = ResolveSchema(\"../schema/gen3_test_schema.json\")\n",
    "resolved_schema.schema\n",
    "# resolved_schema.get_node_link(\"serum_marker_assay.yaml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unhashable type: 'list'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m ref_dict \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m      2\u001b[0m     resolved_schema\u001b[38;5;241m.\u001b[39mschema[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdemographic.yaml\u001b[39m\u001b[38;5;124m'\u001b[39m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlinks\u001b[39m\u001b[38;5;124m'\u001b[39m]: resolved_schema\u001b[38;5;241m.\u001b[39mschema[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdemographic.yaml\u001b[39m\u001b[38;5;124m'\u001b[39m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlinks\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m      3\u001b[0m     }\n",
      "\u001b[0;31mTypeError\u001b[0m: unhashable type: 'list'"
     ]
    }
   ],
   "source": [
    "ref_dict = {\n",
    "    resolved_schema.schema['demographic.yaml']['links']: resolved_schema.schema['demographic.yaml']['links']\n",
    "    }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'backref': 'lipidomics_files',\n",
       "  'label': 'data_from',\n",
       "  'multiplicity': 'many_to_one',\n",
       "  'name': 'lipidomics_assays',\n",
       "  'required': False,\n",
       "  'target_type': 'lipidomics_assay'},\n",
       " {'backref': 'lipidomics_files',\n",
       "  'label': 'data_from',\n",
       "  'multiplicity': 'one_to_one',\n",
       "  'name': 'core_metadata_collections',\n",
       "  'required': False,\n",
       "  'target_type': 'core_metadata_collection'}]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resolved_schema.get_node_link('lipidomics_file.yaml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for getting upstream downstream from links\n",
    "def find_upstream_downstream(links):\n",
    "    \"\"\"Takes a list of dictionaries or a single dictionary of links and returns the upstream and downstream nodes\"\"\"\n",
    "\n",
    "    # Ensure links is a list\n",
    "    if isinstance(links, dict):\n",
    "        links = [links]\n",
    "\n",
    "    results = []\n",
    "\n",
    "    for link in links:\n",
    "        backref = link.get(\"backref\")\n",
    "        target_type = link.get(\"target_type\")\n",
    "        \n",
    "        if not backref or not target_type:\n",
    "            print(\"Missing essential keys in link:\", link)\n",
    "            results.append((None, None))\n",
    "            continue\n",
    "        \n",
    "        # strip last s from name\n",
    "        if backref.endswith('s'):\n",
    "            backref = backref[:-1]\n",
    "\n",
    "        # Determine upstream/downstream logic\n",
    "        upstream = target_type\n",
    "        downstream = backref\n",
    "        \n",
    "        print(f\"Upstream: {upstream}, Downstream: {downstream}\")\n",
    "\n",
    "        results.append((upstream, downstream))\n",
    "\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Upstream: lipidomics_assay, Downstream: lipidomics_file\n",
      "Upstream: core_metadata_collection, Downstream: lipidomics_file\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('lipidomics_assay', 'lipidomics_file'),\n",
       " ('core_metadata_collection', 'lipidomics_file')]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "link = resolved_schema.get_node_link('lipidomics_file.yaml')\n",
    "find_upstream_downstream(link)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
